{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extract 8 regions\n",
    "\n",
    "We're going to extract the 8 regions of interest (ROIs) from the _cis-eQTL_ data. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load packages\n",
    "\n",
    "Here we load the necessary packages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to check for installation of required packages\n",
    "def check_install_package(package_name):\n",
    "    try:\n",
    "        importlib.import_module(package_name)\n",
    "    except ImportError:\n",
    "        print(f'{package_name} is not installed. Install and try again...')\n",
    "        # subprocess.check_call(['pip', 'install', package_name])\n",
    "\n",
    "import os\n",
    "import glob\n",
    "import importlib\n",
    "import subprocess\n",
    "import sys\n",
    "# argument parsing\n",
    "import argparse\n",
    "\n",
    "# get date and time\n",
    "from datetime import datetime\n",
    "\n",
    "# Pandas is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool\n",
    "# check_install_package('pandas')\n",
    "import pandas as pd\n",
    "\n",
    "# pyarrow is supperior to loading parquet files\n",
    "# check_install_package('pyarrow')\n",
    "import pyarrow as pa\n",
    "import pyarrow.parquet as pq\n",
    "\n",
    "# polars is a fast dataframe library\n",
    "# check_install_package('polars')\n",
    "import polars as pl\n",
    "\n",
    "# for statistical analysis\n",
    "# check_install_package('scipy')\n",
    "from scipy import stats\n",
    "import numpy as np\n",
    "\n",
    "# scientific colourmaps\n",
    "# https://www.fabiocrameri.ch/ws/media-library/8c4b111121ff448e843dfef9220bf613/readme_scientificcolourmaps.pdf\n",
    "# check_install_package('cmcrameri')\n",
    "import cmcrameri as ccm\n",
    "from cmcrameri import cm\n",
    "\n",
    "# for plotting\n",
    "# check_install_package('matplotlib')\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "# if using a Jupyter notebook, include:\n",
    "%matplotlib inline\n",
    "\n",
    "# use Seaborn for visualisations\n",
    "# check_install_package('seaborn')\n",
    "import seaborn as sns\n",
    "\n",
    "# for handling GWAS data\n",
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also want some variables set. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking contents of the reference directory:\n",
      "\u001b[34m1000G\u001b[m\u001b[m\n",
      "\u001b[34mGoNL\u001b[m\u001b[m\n",
      "\u001b[34mHRC_r1_1_2016\u001b[m\u001b[m\n",
      "\u001b[34mHRCr11_1000Gp3v5\u001b[m\u001b[m\n",
      "\u001b[34mdbSNP\u001b[m\u001b[m\n",
      "\u001b[34mrefgenie_genomes\u001b[m\u001b[m\n",
      "\u001b[34mtcga\u001b[m\u001b[m\n",
      "\n",
      "Checking contents of the Google Drive directory:\n",
      "\u001b[34mCAC\u001b[m\u001b[m\n",
      "\u001b[34mCIMT\u001b[m\u001b[m\n",
      "\u001b[34mGIGASTROKE\u001b[m\u001b[m\n",
      "\u001b[34mMILLIONHEARTS\u001b[m\u001b[m\n",
      "\u001b[34mPGC\u001b[m\u001b[m\n",
      "\n",
      "Checking contents of the nominal cis-eQTL data directory:\n",
      "README.md\n",
      "tensorqtl_cis_nominal_chr01.cis_qtl_pairs.chr01.parquet\n",
      "tensorqtl_cis_nominal_chr02.cis_qtl_pairs.chr02.parquet\n",
      "tensorqtl_cis_nominal_chr03.cis_qtl_pairs.chr03.parquet\n",
      "tensorqtl_cis_nominal_chr04.cis_qtl_pairs.chr04.parquet\n",
      "tensorqtl_cis_nominal_chr05.cis_qtl_pairs.chr05.parquet\n",
      "tensorqtl_cis_nominal_chr06.cis_qtl_pairs.chr06.parquet\n",
      "tensorqtl_cis_nominal_chr07.cis_qtl_pairs.chr07.parquet\n",
      "tensorqtl_cis_nominal_chr08.cis_qtl_pairs.chr08.parquet\n",
      "tensorqtl_cis_nominal_chr09.cis_qtl_pairs.chr09.parquet\n",
      "tensorqtl_cis_nominal_chr10.cis_qtl_pairs.chr10.parquet\n",
      "tensorqtl_cis_nominal_chr11.cis_qtl_pairs.chr11.parquet\n",
      "tensorqtl_cis_nominal_chr12.cis_qtl_pairs.chr12.parquet\n",
      "tensorqtl_cis_nominal_chr13.cis_qtl_pairs.chr13.parquet\n",
      "tensorqtl_cis_nominal_chr14.cis_qtl_pairs.chr14.parquet\n",
      "tensorqtl_cis_nominal_chr15.cis_qtl_pairs.chr15.parquet\n",
      "tensorqtl_cis_nominal_chr16.cis_qtl_pairs.chr16.parquet\n",
      "tensorqtl_cis_nominal_chr17.cis_qtl_pairs.chr17.parquet\n",
      "tensorqtl_cis_nominal_chr18.cis_qtl_pairs.chr18.parquet\n",
      "tensorqtl_cis_nominal_chr19.cis_qtl_pairs.chr19.parquet\n",
      "tensorqtl_cis_nominal_chr20.cis_qtl_pairs.chr20.parquet\n",
      "tensorqtl_cis_nominal_chr21.cis_qtl_pairs.chr21.parquet\n",
      "tensorqtl_cis_nominal_chr22.cis_qtl_pairs.chr22.parquet\n",
      "tensorqtl_nominal_cis_qtl_pairs.annot.parquet\n",
      "tensorqtl_nominal_cis_qtl_pairs.sumstats.gwaslab\n",
      "tensorqtl_nominal_cis_qtl_pairs.sumstats.log\n",
      "tensorqtl_nominal_cis_qtl_pairs.sumstats.parquet\n",
      "tensorqtl_nominal_cis_qtl_pairs.sumstats.top.csv\n",
      "tensorqtl_nominal_cis_qtl_pairs.sumstats.top.xlsx\n",
      "\n",
      "Checking contents of the permuted cis-mQTL data directory:\n",
      "README.md\n",
      "tensormqtl.perm_cis_mqtl.txt\n",
      "tensormqtl_cis_chr1.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr10.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr11.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr12.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr13.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr14.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr15.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr16.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr17.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr18.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr19.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr2.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr20.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr21.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr22.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr3.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr4.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr5.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr6.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr7.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr8.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr9.cis_qtl.txt.gz\n",
      "\n",
      "Checking contents of the permuted trans-eQTL data directory:\n",
      "README.md\n",
      "tensorqtl_trans_full.trans_qtl_pairs.parquet\n",
      "\n",
      "Checking contents of the permuted trans-eQTL data directory:\n",
      "README.md\n",
      "tensormqtl_perm_trans_qtl_pairs.annot.parquet\n",
      "tensormqtl_trans_chr1.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr10.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr11.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr12.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr13.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr14.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr15.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr16.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr17.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr18.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr19.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr2.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr20.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr21.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr22.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr3.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr4.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr5.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr6.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr7.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr8.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr9.trans_qtl_pairs.parquet\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Create directories for the GWAS data and the reference data\n",
    "import os\n",
    "from subprocess import check_output\n",
    "\n",
    "# set some general defaults\n",
    "PGC_HITS = \"PGC\"\n",
    "\n",
    "POPULATION = \"EUR\"\n",
    "\n",
    "# general plotting directory\n",
    "PLOTS_loc = \"PLOTS\"\n",
    "\n",
    "# location to put molQTL results\n",
    "molQTL_loc = \"molQTL_results\"\n",
    "\n",
    "# Check if the directory exists\n",
    "if not os.path.exists(molQTL_loc):\n",
    "    # If it doesn't exist, create it\n",
    "    os.makedirs(molQTL_loc)\n",
    "\n",
    "# Check if the directory exists\n",
    "if not os.path.exists(PLOTS_loc):\n",
    "    # If it doesn't exist, create it\n",
    "    os.makedirs(PLOTS_loc)\n",
    "\n",
    "# # regional association plots directory\n",
    "# REG_PLOTS_loc = PLOTS_loc + \"/Regional_Association_Plots\"\n",
    "\n",
    "# # Check if the directory exists\n",
    "# if not os.path.exists(REG_PLOTS_loc):\n",
    "#     # If it doesn't exist, create it\n",
    "#     os.makedirs(REG_PLOTS_loc)\n",
    "\n",
    "# Reference data directory\n",
    "REF_loc = \"/Users/slaan3/PLINK/references\"\n",
    "print(\"Checking contents of the reference directory:\")\n",
    "print(check_output([\"ls\", os.path.join(REF_loc)]).decode(\"utf8\"))\n",
    "\n",
    "# GWAS data directory\n",
    "GD_loc = \"/Users/slaan3/Library/CloudStorage/GoogleDrive-s.w.vanderlaan@gmail.com/My Drive/Genomics/#Projects/TO_AITION/MR CVD MDD/GWAS\"\n",
    "print(\"Checking contents of the Google Drive directory:\")\n",
    "print(check_output([\"ls\", os.path.join(GD_loc)]).decode(\"utf8\"))\n",
    "\n",
    "# molQTL data directory\n",
    "NOM_CIS_EQTL_loc = \"/Users/slaan3/git/CirculatoryHealth/molqtl/results/version1_aernas1_firstrun/nom_cis_eqtl\"\n",
    "print(\"Checking contents of the nominal cis-eQTL data directory:\")\n",
    "print(check_output([\"ls\", os.path.join(NOM_CIS_EQTL_loc)]).decode(\"utf8\"))\n",
    "\n",
    "PERM_CIS_MQTL_loc = \"/Users/slaan3/git/CirculatoryHealth/molqtl/results/perm_cis_mqtl\"\n",
    "print(\"Checking contents of the permuted cis-mQTL data directory:\")\n",
    "print(check_output([\"ls\", os.path.join(PERM_CIS_MQTL_loc)]).decode(\"utf8\"))\n",
    "\n",
    "PERM_TRANS_EQTL_loc = \"/Users/slaan3/git/CirculatoryHealth/molqtl/results/version1_aernas1_firstrun/perm_trans_eqtl\"\n",
    "print(\"Checking contents of the permuted trans-eQTL data directory:\")\n",
    "print(check_output([\"ls\", os.path.join(PERM_TRANS_EQTL_loc)]).decode(\"utf8\"))\n",
    "\n",
    "PERM_TRANS_MQTL_loc = (\n",
    "    \"/Users/slaan3/git/CirculatoryHealth/molqtl/results/perm_trans_mqtl\"\n",
    ")\n",
    "print(\"Checking contents of the permuted trans-eQTL data directory:\")\n",
    "print(check_output([\"ls\", os.path.join(PERM_TRANS_MQTL_loc)]).decode(\"utf8\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data\n",
    "\n",
    "We are going to need the list of ROIs and the annotated cis-eQTL data."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Regions-of-interest\n",
    "\n",
    "Let's load the 8 regions we are interested in."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# polars.read_excel(\n",
    "# source: str | BytesIO | Path | BinaryIO | bytes,\n",
    "# *,\n",
    "# sheet_id: None = None,\n",
    "# sheet_name: str,\n",
    "# engine: Literal['xlsx2csv', 'openpyxl', 'pyxlsb'] | None = None,\n",
    "# xlsx2csv_options: dict[str, Any] | None = None,\n",
    "# read_csv_options: dict[str, Any] | None = None,\n",
    "# schema_overrides: SchemaDict | None = None,\n",
    "# raise_if_empty: bool = True,\n",
    "# )\n",
    "\n",
    "target_regions = pl.read_csv(\n",
    "    source=os.path.join(\"targets/regions.csv\")\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (8, 4)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>leadSNP</th><th>chromosome</th><th>lower_position</th><th>upper_position</th></tr><tr><td>str</td><td>i64</td><td>i64</td><td>i64</td></tr></thead><tbody><tr><td>&quot;7:12263538&quot;</td><td>7</td><td>11263538</td><td>13263538</td></tr><tr><td>&quot;19:4044579&quot;</td><td>19</td><td>3044579</td><td>5044579</td></tr><tr><td>&quot;15:38843887&quot;</td><td>15</td><td>37843887</td><td>39843887</td></tr><tr><td>&quot;7:114059156&quot;</td><td>7</td><td>113059156</td><td>115059156</td></tr><tr><td>&quot;7:2760750&quot;</td><td>7</td><td>1760750</td><td>3760750</td></tr><tr><td>&quot;7:1937261&quot;</td><td>7</td><td>937261</td><td>2937261</td></tr><tr><td>&quot;2:165045101&quot;</td><td>2</td><td>164045101</td><td>166045101</td></tr><tr><td>&quot;2:164915279&quot;</td><td>2</td><td>163915279</td><td>165915279</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (8, 4)\n",
       "┌─────────────┬────────────┬────────────────┬────────────────┐\n",
       "│ leadSNP     ┆ chromosome ┆ lower_position ┆ upper_position │\n",
       "│ ---         ┆ ---        ┆ ---            ┆ ---            │\n",
       "│ str         ┆ i64        ┆ i64            ┆ i64            │\n",
       "╞═════════════╪════════════╪════════════════╪════════════════╡\n",
       "│ 7:12263538  ┆ 7          ┆ 11263538       ┆ 13263538       │\n",
       "│ 19:4044579  ┆ 19         ┆ 3044579        ┆ 5044579        │\n",
       "│ 15:38843887 ┆ 15         ┆ 37843887       ┆ 39843887       │\n",
       "│ 7:114059156 ┆ 7          ┆ 113059156      ┆ 115059156      │\n",
       "│ 7:2760750   ┆ 7          ┆ 1760750        ┆ 3760750        │\n",
       "│ 7:1937261   ┆ 7          ┆ 937261         ┆ 2937261        │\n",
       "│ 2:165045101 ┆ 2          ┆ 164045101      ┆ 166045101      │\n",
       "│ 2:164915279 ┆ 2          ┆ 163915279      ┆ 165915279      │\n",
       "└─────────────┴────────────┴────────────────┴────────────────┘"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_regions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### cis-eQTL data\n",
    "\n",
    "We previously parsed the cis-eQTL data into a `.parquet` file which includes annotated results. For details check `` in the `` GitHub repository. \n",
    "\n",
    "Here we'll simply load thed annotated data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in data\n",
    "# https://stackoverflow.com/questions/33813815/how-to-read-a-parquet-file-into-pandas-dataframe\n",
    "\n",
    "import polars as pl\n",
    "\n",
    "# annotated cis-eQTLs\n",
    "sumstats_nom_cis_eqtl = pl.read_parquet(\n",
    "    source=os.path.join(\n",
    "        NOM_CIS_EQTL_loc, \"tensorqtl_nominal_cis_qtl_pairs.annot.parquet\"\n",
    "    )\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (77_285_276, 25)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>EnsemblID</th><th>VariantID</th><th>tss_distance</th><th>CAF_eQTL</th><th>ma_samples</th><th>ma_count</th><th>pval_nominal</th><th>Beta</th><th>SE</th><th>chromosome</th><th>position</th><th>OtherAlleleA</th><th>CodedAlleleB</th><th>CAF</th><th>AltID</th><th>Source</th><th>AverageMaximumPosteriorCall</th><th>Info</th><th>AA_N</th><th>AB_N</th><th>BB_N</th><th>TotalN</th><th>MAF</th><th>MissingDataProportion</th><th>HWE_P</th></tr><tr><td>str</td><td>str</td><td>i32</td><td>f32</td><td>i32</td><td>i32</td><td>f64</td><td>f32</td><td>f32</td><td>i64</td><td>i64</td><td>str</td><td>str</td><td>f64</td><td>str</td><td>str</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>i64</td><td>f64</td><td>f64</td><td>f64</td></tr></thead><tbody><tr><td>&quot;ENSG00000187634&quot;</td><td>&quot;1:693731&quot;</td><td>-240519</td><td>0.134185</td><td>153</td><td>168</td><td>0.881492</td><td>0.005953</td><td>0.039917</td><td>1</td><td>693731</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.137241</td><td>&quot;rs12238997&quot;</td><td>&quot;HRCr11&quot;</td><td>0.902046</td><td>0.624501</td><td>1637.01</td><td>451.103</td><td>35.569</td><td>2124</td><td>0.122957</td><td>0.000075</td><td>0.544841</td></tr><tr><td>&quot;ENSG00000187634&quot;</td><td>&quot;1:714596&quot;</td><td>-219654</td><td>0.038339</td><td>48</td><td>48</td><td>0.834746</td><td>-0.015369</td><td>0.073635</td><td>1</td><td>714596</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.0327213</td><td>&quot;rs149887893&quot;</td><td>&quot;HRCr11&quot;</td><td>0.968278</td><td>0.577564</td><td>1987.73</td><td>135.495</td><td>0.677</td><td>2124</td><td>0.0322165</td><td>0.000024</td><td>0.272504</td></tr><tr><td>&quot;ENSG00000187634&quot;</td><td>&quot;1:715367&quot;</td><td>-218883</td><td>0.038339</td><td>48</td><td>48</td><td>0.834746</td><td>-0.015369</td><td>0.073635</td><td>1</td><td>715367</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.0324859</td><td>&quot;rs12184277&quot;</td><td>&quot;HRCr11&quot;</td><td>0.975207</td><td>0.67196</td><td>1976.85</td><td>146.512</td><td>0.556</td><td>2124</td><td>0.0347528</td><td>0.000019</td><td>0.176551</td></tr><tr><td>&quot;ENSG00000187634&quot;</td><td>&quot;1:717485&quot;</td><td>-216765</td><td>0.038339</td><td>48</td><td>48</td><td>0.834746</td><td>-0.015369</td><td>0.073635</td><td>1</td><td>717485</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>0.0324859</td><td>&quot;rs12184279&quot;</td><td>&quot;HRCr11&quot;</td><td>0.975183</td><td>0.670147</td><td>1978.35</td><td>145.041</td><td>0.517</td><td>2124</td><td>0.0343883</td><td>0.000022</td><td>0.174817</td></tr><tr><td>&quot;ENSG00000187634&quot;</td><td>&quot;1:720381&quot;</td><td>-213869</td><td>0.038339</td><td>48</td><td>48</td><td>0.834746</td><td>-0.015369</td><td>0.073635</td><td>1</td><td>720381</td><td>&quot;G&quot;</td><td>&quot;T&quot;</td><td>0.0327213</td><td>&quot;rs116801199&quot;</td><td>&quot;HRCr11&quot;</td><td>0.974476</td><td>0.667536</td><td>1973.72</td><td>149.503</td><td>0.691</td><td>2124</td><td>0.0355205</td><td>0.00002</td><td>0.110452</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;ENSG00000079974&quot;</td><td>&quot;22:51220249&quot;</td><td>444664</td><td>0.035942</td><td>43</td><td>45</td><td>0.863512</td><td>-0.004492</td><td>0.026119</td><td>22</td><td>51220249</td><td>&quot;A&quot;</td><td>&quot;T&quot;</td><td>0.0442561</td><td>&quot;rs9616977&quot;</td><td>&quot;1000Gp3v5&quot;</td><td>0.913151</td><td>0.403798</td><td>1832.6</td><td>278.562</td><td>12.692</td><td>2124</td><td>0.0715553</td><td>0.000034</td><td>0.622809</td></tr><tr><td>&quot;ENSG00000079974&quot;</td><td>&quot;22:51221190&quot;</td><td>445605</td><td>0.035144</td><td>42</td><td>44</td><td>0.671267</td><td>0.011112</td><td>0.026169</td><td>22</td><td>51221190</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>0.0400188</td><td>&quot;rs369304721&quot;</td><td>&quot;HRCr11&quot;</td><td>0.933398</td><td>0.443438</td><td>1911.83</td><td>204.73</td><td>7.373</td><td>2124</td><td>0.0516673</td><td>0.000015</td><td>0.500544</td></tr><tr><td>&quot;ENSG00000079974&quot;</td><td>&quot;22:51221731&quot;</td><td>446146</td><td>0.035942</td><td>43</td><td>45</td><td>0.863512</td><td>-0.004492</td><td>0.026119</td><td>22</td><td>51221731</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.0442561</td><td>&quot;rs115055839&quot;</td><td>&quot;HRCr11&quot;</td><td>0.911894</td><td>0.405944</td><td>1826.36</td><td>284.126</td><td>13.448</td><td>2124</td><td>0.0732185</td><td>0.000016</td><td>0.523538</td></tr><tr><td>&quot;ENSG00000079974&quot;</td><td>&quot;22:51222100&quot;</td><td>446515</td><td>0.054313</td><td>67</td><td>68</td><td>0.408353</td><td>0.018264</td><td>0.022074</td><td>22</td><td>51222100</td><td>&quot;G&quot;</td><td>&quot;T&quot;</td><td>0.0503766</td><td>&quot;rs114553188&quot;</td><td>&quot;HRCr11&quot;</td><td>0.969267</td><td>0.716716</td><td>1911.95</td><td>205.392</td><td>6.52</td><td>2124</td><td>0.0514234</td><td>0.000033</td><td>0.821168</td></tr><tr><td>&quot;ENSG00000079974&quot;</td><td>&quot;22:51229805&quot;</td><td>454220</td><td>0.038339</td><td>46</td><td>48</td><td>0.973919</td><td>-0.000827</td><td>0.025284</td><td>22</td><td>51229805</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.0477872</td><td>&quot;rs9616985&quot;</td><td>&quot;HRCr11&quot;</td><td>0.935419</td><td>0.585</td><td>1821.25</td><td>287.583</td><td>15.116</td><td>2124</td><td>0.0748171</td><td>0.000012</td><td>0.342973</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (77_285_276, 25)\n",
       "┌────────────┬────────────┬────────────┬──────────┬───┬────────┬───────────┬────────────┬──────────┐\n",
       "│ EnsemblID  ┆ VariantID  ┆ tss_distan ┆ CAF_eQTL ┆ … ┆ TotalN ┆ MAF       ┆ MissingDat ┆ HWE_P    │\n",
       "│ ---        ┆ ---        ┆ ce         ┆ ---      ┆   ┆ ---    ┆ ---       ┆ aProportio ┆ ---      │\n",
       "│ str        ┆ str        ┆ ---        ┆ f32      ┆   ┆ i64    ┆ f64       ┆ n          ┆ f64      │\n",
       "│            ┆            ┆ i32        ┆          ┆   ┆        ┆           ┆ ---        ┆          │\n",
       "│            ┆            ┆            ┆          ┆   ┆        ┆           ┆ f64        ┆          │\n",
       "╞════════════╪════════════╪════════════╪══════════╪═══╪════════╪═══════════╪════════════╪══════════╡\n",
       "│ ENSG000001 ┆ 1:693731   ┆ -240519    ┆ 0.134185 ┆ … ┆ 2124   ┆ 0.122957  ┆ 0.000075   ┆ 0.544841 │\n",
       "│ 87634      ┆            ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
       "│ ENSG000001 ┆ 1:714596   ┆ -219654    ┆ 0.038339 ┆ … ┆ 2124   ┆ 0.0322165 ┆ 0.000024   ┆ 0.272504 │\n",
       "│ 87634      ┆            ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
       "│ ENSG000001 ┆ 1:715367   ┆ -218883    ┆ 0.038339 ┆ … ┆ 2124   ┆ 0.0347528 ┆ 0.000019   ┆ 0.176551 │\n",
       "│ 87634      ┆            ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
       "│ ENSG000001 ┆ 1:717485   ┆ -216765    ┆ 0.038339 ┆ … ┆ 2124   ┆ 0.0343883 ┆ 0.000022   ┆ 0.174817 │\n",
       "│ 87634      ┆            ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
       "│ ENSG000001 ┆ 1:720381   ┆ -213869    ┆ 0.038339 ┆ … ┆ 2124   ┆ 0.0355205 ┆ 0.00002    ┆ 0.110452 │\n",
       "│ 87634      ┆            ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
       "│ …          ┆ …          ┆ …          ┆ …        ┆ … ┆ …      ┆ …         ┆ …          ┆ …        │\n",
       "│ ENSG000000 ┆ 22:5122024 ┆ 444664     ┆ 0.035942 ┆ … ┆ 2124   ┆ 0.0715553 ┆ 0.000034   ┆ 0.622809 │\n",
       "│ 79974      ┆ 9          ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
       "│ ENSG000000 ┆ 22:5122119 ┆ 445605     ┆ 0.035144 ┆ … ┆ 2124   ┆ 0.0516673 ┆ 0.000015   ┆ 0.500544 │\n",
       "│ 79974      ┆ 0          ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
       "│ ENSG000000 ┆ 22:5122173 ┆ 446146     ┆ 0.035942 ┆ … ┆ 2124   ┆ 0.0732185 ┆ 0.000016   ┆ 0.523538 │\n",
       "│ 79974      ┆ 1          ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
       "│ ENSG000000 ┆ 22:5122210 ┆ 446515     ┆ 0.054313 ┆ … ┆ 2124   ┆ 0.0514234 ┆ 0.000033   ┆ 0.821168 │\n",
       "│ 79974      ┆ 0          ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
       "│ ENSG000000 ┆ 22:5122980 ┆ 454220     ┆ 0.038339 ┆ … ┆ 2124   ┆ 0.0748171 ┆ 0.000012   ┆ 0.342973 │\n",
       "│ 79974      ┆ 5          ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
       "└────────────┴────────────┴────────────┴──────────┴───┴────────┴───────────┴────────────┴──────────┘"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sumstats_nom_cis_eqtl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract data\n",
    "\n",
    "Now we're ready to extract the data for each region in one table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape: (621_506, 25)\n",
      "┌────────────┬────────────┬────────────┬──────────┬───┬────────┬───────────┬────────────┬──────────┐\n",
      "│ EnsemblID  ┆ VariantID  ┆ tss_distan ┆ CAF_eQTL ┆ … ┆ TotalN ┆ MAF       ┆ MissingDat ┆ HWE_P    │\n",
      "│ ---        ┆ ---        ┆ ce         ┆ ---      ┆   ┆ ---    ┆ ---       ┆ aProportio ┆ ---      │\n",
      "│ str        ┆ str        ┆ ---        ┆ f32      ┆   ┆ i64    ┆ f64       ┆ n          ┆ f64      │\n",
      "│            ┆            ┆ i32        ┆          ┆   ┆        ┆           ┆ ---        ┆          │\n",
      "│            ┆            ┆            ┆          ┆   ┆        ┆           ┆ f64        ┆          │\n",
      "╞════════════╪════════════╪════════════╪══════════╪═══╪════════╪═══════════╪════════════╪══════════╡\n",
      "│ ENSG000001 ┆ 7:11263538 ┆ 327489     ┆ 0.071086 ┆ … ┆ 2124   ┆ 0.0684027 ┆ 0.000001   ┆ 0.610073 │\n",
      "│ 89043      ┆            ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
      "│ ENSG000001 ┆ 7:11264536 ┆ 328487     ┆ 0.047923 ┆ … ┆ 2124   ┆ 0.0536334 ┆ 0.000004   ┆ 0.830364 │\n",
      "│ 89043      ┆            ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
      "│ ENSG000001 ┆ 7:11264670 ┆ 328621     ┆ 0.075879 ┆ … ┆ 2124   ┆ 0.0805431 ┆ 0.000006   ┆ 0.304873 │\n",
      "│ 89043      ┆            ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
      "│ ENSG000001 ┆ 7:11264692 ┆ 328643     ┆ 0.521565 ┆ … ┆ 2124   ┆ 0.45498   ┆ 0.000005   ┆ 1.0      │\n",
      "│ 89043      ┆            ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
      "│ ENSG000001 ┆ 7:11265421 ┆ 329372     ┆ 0.598243 ┆ … ┆ 2124   ┆ 0.38279   ┆ 0.000004   ┆ 0.408573 │\n",
      "│ 89043      ┆            ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
      "│ …          ┆ …          ┆ …          ┆ …        ┆ … ┆ …      ┆ …         ┆ …          ┆ …        │\n",
      "│ ENSG000001 ┆ 2:16591356 ┆ -593964    ┆ 0.646965 ┆ … ┆ 2124   ┆ 0.350612  ┆ 0.000003   ┆ 0.390789 │\n",
      "│ 36546      ┆ 4          ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
      "│ ENSG000001 ┆ 2:16591424 ┆ -593287    ┆ 0.162939 ┆ … ┆ 2124   ┆ 0.17201   ┆ 0.000009   ┆ 0.322517 │\n",
      "│ 36546      ┆ 1          ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
      "│ ENSG000001 ┆ 2:16591435 ┆ -593175    ┆ 0.227636 ┆ … ┆ 2124   ┆ 0.226552  ┆ 0.000008   ┆ 0.292141 │\n",
      "│ 36546      ┆ 3          ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
      "│ ENSG000001 ┆ 2:16591441 ┆ -593112    ┆ 0.234026 ┆ … ┆ 2124   ┆ 0.234516  ┆ 0.000008   ┆ 0.671834 │\n",
      "│ 36546      ┆ 6          ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
      "│ ENSG000001 ┆ 2:16591480 ┆ -592726    ┆ 0.13738  ┆ … ┆ 2124   ┆ 0.137259  ┆ 0.000004   ┆ 0.141703 │\n",
      "│ 36546      ┆ 2          ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
      "└────────────┴────────────┴────────────┴──────────┴───┴────────┴───────────┴────────────┴──────────┘\n"
     ]
    }
   ],
   "source": [
    "import polars as pl\n",
    "\n",
    "# Initialize an empty DataFrame to store the filtered results\n",
    "filtered_results = pl.DataFrame()\n",
    "\n",
    "# Loop through each row of df2 and filter df1 based on chromosome and position range\n",
    "for row in target_regions.iter_rows(named=True):\n",
    "    chrom = row['chromosome']\n",
    "    lower_pos = row['lower_position']\n",
    "    upper_pos = row['upper_position']\n",
    "    \n",
    "    # Filter df1 based on the chromosome and position range using polars\n",
    "    filtered = sumstats_nom_cis_eqtl.filter(\n",
    "        (sumstats_nom_cis_eqtl['chromosome'] == chrom) & \n",
    "        (sumstats_nom_cis_eqtl['position'] >= lower_pos) & \n",
    "        (sumstats_nom_cis_eqtl['position'] <= upper_pos)\n",
    "    )\n",
    "    \n",
    "    # Append the filtered rows to the result DataFrame\n",
    "    filtered_results = filtered_results.vstack(filtered)\n",
    "\n",
    "# Show the final filtered DataFrame\n",
    "print(filtered_results)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows in filtered_results: 621506\n"
     ]
    }
   ],
   "source": [
    "# Count the number of rows in the filtered_results DataFrame\n",
    "num_rows = filtered_results.shape[0]\n",
    "\n",
    "# Print the number of rows\n",
    "print(f\"Number of rows in filtered_results: {num_rows}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "del sumstats_perm_trans_mqtl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotation\n",
    "\n",
    "Let's also annotate this table with more information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:biothings.client:querying 1-176...\n",
      "INFO:biothings.client:done.\n",
      "INFO:biothings.client:Finished.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>alias</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>map_location</th>\n",
       "      <th>genomic_pos</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>HGNC</th>\n",
       "      <th>summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000140320</td>\n",
       "      <td>BAHD1</td>\n",
       "      <td>bromo adjacent homology domain containing 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>15q15.1</td>\n",
       "      <td>{'chr': '15', 'end': 40468236, 'ensemblgene': ...</td>\n",
       "      <td>22893</td>\n",
       "      <td>29153</td>\n",
       "      <td>Enables chromatin binding activity. Involved i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000064961</td>\n",
       "      <td>HMG20B</td>\n",
       "      <td>high mobility group 20B</td>\n",
       "      <td>[BRAF25, BRAF35, HMGX2, HMGXB2, PP7706, SMARCE...</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>19p13.3</td>\n",
       "      <td>{'chr': '19', 'end': 3579088, 'ensemblgene': '...</td>\n",
       "      <td>10362</td>\n",
       "      <td>5002</td>\n",
       "      <td>Predicted to enable DNA binding activity. Pred...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000065717</td>\n",
       "      <td>TLE2</td>\n",
       "      <td>TLE family member 2, transcriptional corepressor</td>\n",
       "      <td>[ESG, ESG2, GRG2]</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>19p13.3</td>\n",
       "      <td>{'chr': '19', 'end': 3047635, 'ensemblgene': '...</td>\n",
       "      <td>7089</td>\n",
       "      <td>11838</td>\n",
       "      <td>Enables transcription corepressor activity. In...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000146802</td>\n",
       "      <td>TMEM168</td>\n",
       "      <td>transmembrane protein 168</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>7q31.1</td>\n",
       "      <td>{'chr': '7', 'end': 112790423, 'ensemblgene': ...</td>\n",
       "      <td>64418</td>\n",
       "      <td>25826</td>\n",
       "      <td>Located in transport vesicle. [provided by All...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000073067</td>\n",
       "      <td>CYP2W1</td>\n",
       "      <td>cytochrome P450 family 2 subfamily W member 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>7p22.3</td>\n",
       "      <td>{'chr': '7', 'end': 989640, 'ensemblgene': 'EN...</td>\n",
       "      <td>54905</td>\n",
       "      <td>20243</td>\n",
       "      <td>This gene encodes a member of the cytochrome P...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>ENSG00000166068</td>\n",
       "      <td>SPRED1</td>\n",
       "      <td>sprouty related EVH1 domain containing 1</td>\n",
       "      <td>[LGSS, NFLS, PPP1R147, hSpred1, spred-1]</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>15q14</td>\n",
       "      <td>{'chr': '15', 'end': 38357249, 'ensemblgene': ...</td>\n",
       "      <td>161742</td>\n",
       "      <td>20249</td>\n",
       "      <td>The protein encoded by this gene is a member o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>ENSG00000130255</td>\n",
       "      <td>RPL36</td>\n",
       "      <td>ribosomal protein L36</td>\n",
       "      <td>[L36, eL36]</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>19p13.3</td>\n",
       "      <td>{'chr': '19', 'end': 5691875, 'ensemblgene': '...</td>\n",
       "      <td>25873</td>\n",
       "      <td>13631</td>\n",
       "      <td>Ribosomes, the organelles that catalyze protei...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>ENSG00000105229</td>\n",
       "      <td>PIAS4</td>\n",
       "      <td>protein inhibitor of activated STAT 4</td>\n",
       "      <td>[PIAS-gamma, PIASY, Piasg, ZMIZ6]</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>19p13.3</td>\n",
       "      <td>{'chr': '19', 'end': 4039386, 'ensemblgene': '...</td>\n",
       "      <td>51588</td>\n",
       "      <td>17002</td>\n",
       "      <td>Enables SUMO ligase activity and ubiquitin pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>ENSG00000128891</td>\n",
       "      <td>CCDC32</td>\n",
       "      <td>coiled-coil domain containing 32</td>\n",
       "      <td>[C15orf57, CFNDS]</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>15q15.1</td>\n",
       "      <td>{'chr': '15', 'end': 40565054, 'ensemblgene': ...</td>\n",
       "      <td>90416</td>\n",
       "      <td>28295</td>\n",
       "      <td>Involved in head development. [provided by All...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>ENSG00000176619</td>\n",
       "      <td>LMNB2</td>\n",
       "      <td>lamin B2</td>\n",
       "      <td>[EPM9, LAMB2, LMN2, MCPH27]</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>19p13.3</td>\n",
       "      <td>{'chr': '19', 'end': 2456959, 'ensemblgene': '...</td>\n",
       "      <td>84823</td>\n",
       "      <td>6638</td>\n",
       "      <td>This gene encodes a B type nuclear lamin. The ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>176 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               query   symbol  \\\n",
       "0    ENSG00000140320    BAHD1   \n",
       "1    ENSG00000064961   HMG20B   \n",
       "2    ENSG00000065717     TLE2   \n",
       "3    ENSG00000146802  TMEM168   \n",
       "4    ENSG00000073067   CYP2W1   \n",
       "..               ...      ...   \n",
       "171  ENSG00000166068   SPRED1   \n",
       "172  ENSG00000130255    RPL36   \n",
       "173  ENSG00000105229    PIAS4   \n",
       "174  ENSG00000128891   CCDC32   \n",
       "175  ENSG00000176619    LMNB2   \n",
       "\n",
       "                                                 name  \\\n",
       "0         bromo adjacent homology domain containing 1   \n",
       "1                             high mobility group 20B   \n",
       "2    TLE family member 2, transcriptional corepressor   \n",
       "3                           transmembrane protein 168   \n",
       "4       cytochrome P450 family 2 subfamily W member 1   \n",
       "..                                                ...   \n",
       "171          sprouty related EVH1 domain containing 1   \n",
       "172                             ribosomal protein L36   \n",
       "173             protein inhibitor of activated STAT 4   \n",
       "174                  coiled-coil domain containing 32   \n",
       "175                                          lamin B2   \n",
       "\n",
       "                                                 alias    type_of_gene  \\\n",
       "0                                                  NaN  protein-coding   \n",
       "1    [BRAF25, BRAF35, HMGX2, HMGXB2, PP7706, SMARCE...  protein-coding   \n",
       "2                                    [ESG, ESG2, GRG2]  protein-coding   \n",
       "3                                                  NaN  protein-coding   \n",
       "4                                                  NaN  protein-coding   \n",
       "..                                                 ...             ...   \n",
       "171           [LGSS, NFLS, PPP1R147, hSpred1, spred-1]  protein-coding   \n",
       "172                                        [L36, eL36]  protein-coding   \n",
       "173                  [PIAS-gamma, PIASY, Piasg, ZMIZ6]  protein-coding   \n",
       "174                                  [C15orf57, CFNDS]  protein-coding   \n",
       "175                        [EPM9, LAMB2, LMN2, MCPH27]  protein-coding   \n",
       "\n",
       "    map_location                                        genomic_pos  \\\n",
       "0        15q15.1  {'chr': '15', 'end': 40468236, 'ensemblgene': ...   \n",
       "1        19p13.3  {'chr': '19', 'end': 3579088, 'ensemblgene': '...   \n",
       "2        19p13.3  {'chr': '19', 'end': 3047635, 'ensemblgene': '...   \n",
       "3         7q31.1  {'chr': '7', 'end': 112790423, 'ensemblgene': ...   \n",
       "4         7p22.3  {'chr': '7', 'end': 989640, 'ensemblgene': 'EN...   \n",
       "..           ...                                                ...   \n",
       "171        15q14  {'chr': '15', 'end': 38357249, 'ensemblgene': ...   \n",
       "172      19p13.3  {'chr': '19', 'end': 5691875, 'ensemblgene': '...   \n",
       "173      19p13.3  {'chr': '19', 'end': 4039386, 'ensemblgene': '...   \n",
       "174      15q15.1  {'chr': '15', 'end': 40565054, 'ensemblgene': ...   \n",
       "175      19p13.3  {'chr': '19', 'end': 2456959, 'ensemblgene': '...   \n",
       "\n",
       "    entrezgene   HGNC                                            summary  \n",
       "0        22893  29153  Enables chromatin binding activity. Involved i...  \n",
       "1        10362   5002  Predicted to enable DNA binding activity. Pred...  \n",
       "2         7089  11838  Enables transcription corepressor activity. In...  \n",
       "3        64418  25826  Located in transport vesicle. [provided by All...  \n",
       "4        54905  20243  This gene encodes a member of the cytochrome P...  \n",
       "..         ...    ...                                                ...  \n",
       "171     161742  20249  The protein encoded by this gene is a member o...  \n",
       "172      25873  13631  Ribosomes, the organelles that catalyze protei...  \n",
       "173      51588  17002  Enables SUMO ligase activity and ubiquitin pro...  \n",
       "174      90416  28295  Involved in head development. [provided by All...  \n",
       "175      84823   6638  This gene encodes a B type nuclear lamin. The ...  \n",
       "\n",
       "[176 rows x 10 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import mygene\n",
    "import pandas as pd\n",
    "import polars as pl\n",
    "\n",
    "# Initialize mygene client\n",
    "mg = mygene.MyGeneInfo()\n",
    "\n",
    "# Convert the polars DataFrame (filtered_results) to a pandas DataFrame\n",
    "filtered_results_pd = filtered_results.to_pandas()\n",
    "\n",
    "# Get the unique Ensembl IDs from the polars DataFrame (filtered_results)\n",
    "ensembl_ids = filtered_results['EnsemblID'].unique().to_list()\n",
    "\n",
    "# Query mygene to get additional gene information (e.g., symbol, name, etc.)\n",
    "gene_info = mg.querymany(ensembl_ids, \n",
    "                         scopes='ensembl.gene', \n",
    "                         fields='symbol,name,alias,type_of_gene,map_location,genomic_pos,entrezgene,HGNC,summary', \n",
    "                         species='human')\n",
    "\n",
    "# Convert the gene info to a pandas DataFrame\n",
    "gene_info_df = pd.DataFrame(gene_info)\n",
    "\n",
    "# Select relevant columns\n",
    "gene_info_df = gene_info_df[['query', 'symbol', 'name', 'alias', 'type_of_gene', 'map_location', 'genomic_pos', 'entrezgene', 'HGNC', 'summary']]\n",
    "\n",
    "# Convert the pandas DataFrame back to polars DataFrame\n",
    "gene_info_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              EnsemblID    VariantID  tss_distance  CAF_eQTL  ma_samples  \\\n",
      "0       ENSG00000189043   7:11263538        327489  0.071086          84   \n",
      "1       ENSG00000189043   7:11264536        328487  0.047923          56   \n",
      "2       ENSG00000189043   7:11264670        328621  0.075879          88   \n",
      "3       ENSG00000189043   7:11264692        328643  0.521565         455   \n",
      "4       ENSG00000189043   7:11265421        329372  0.598243         404   \n",
      "...                 ...          ...           ...       ...         ...   \n",
      "621501  ENSG00000136546  2:165913564       -593964  0.646965         362   \n",
      "621502  ENSG00000136546  2:165914241       -593287  0.162939         186   \n",
      "621503  ENSG00000136546  2:165914353       -593175  0.227636         251   \n",
      "621504  ENSG00000136546  2:165914416       -593112  0.234026         259   \n",
      "621505  ENSG00000136546  2:165914802       -592726  0.137380         155   \n",
      "\n",
      "        ma_count  pval_nominal      Beta        SE  chromosome  ...     HWE_P  \\\n",
      "0             89      0.337176 -0.027423  0.028549           7  ...  0.610073   \n",
      "1             60      0.349662  0.032065  0.034257           7  ...  0.830364   \n",
      "2             95      0.299686 -0.028301  0.027264           7  ...  0.304873   \n",
      "3            599      0.789388 -0.004011  0.015009           7  ...  1.000000   \n",
      "4            503      0.354785 -0.014423  0.015574           7  ...  0.408573   \n",
      "...          ...           ...       ...       ...         ...  ...       ...   \n",
      "621501       442      0.026601  0.039790  0.017899           2  ...  0.390789   \n",
      "621502       204      0.591413 -0.012557  0.023379           2  ...  0.322517   \n",
      "621503       285      0.338273  0.019565  0.020414           2  ...  0.292141   \n",
      "621504       293      0.031736  0.043972  0.020424           2  ...  0.671834   \n",
      "621505       172      0.873595 -0.003890  0.024438           2  ...  0.141703   \n",
      "\n",
      "        symbol                                          name  \\\n",
      "0       NDUFA4       NDUFA4 mitochondrial complex associated   \n",
      "1       NDUFA4       NDUFA4 mitochondrial complex associated   \n",
      "2       NDUFA4       NDUFA4 mitochondrial complex associated   \n",
      "3       NDUFA4       NDUFA4 mitochondrial complex associated   \n",
      "4       NDUFA4       NDUFA4 mitochondrial complex associated   \n",
      "...        ...                                           ...   \n",
      "621501   SCN7A  sodium voltage-gated channel alpha subunit 7   \n",
      "621502   SCN7A  sodium voltage-gated channel alpha subunit 7   \n",
      "621503   SCN7A  sodium voltage-gated channel alpha subunit 7   \n",
      "621504   SCN7A  sodium voltage-gated channel alpha subunit 7   \n",
      "621505   SCN7A  sodium voltage-gated channel alpha subunit 7   \n",
      "\n",
      "                                                    alias    type_of_gene  \\\n",
      "0       [CI-9k, CI-MLRQ, COXFA4, MC4DN21, MISTR1, MLRQ...  protein-coding   \n",
      "1       [CI-9k, CI-MLRQ, COXFA4, MC4DN21, MISTR1, MLRQ...  protein-coding   \n",
      "2       [CI-9k, CI-MLRQ, COXFA4, MC4DN21, MISTR1, MLRQ...  protein-coding   \n",
      "3       [CI-9k, CI-MLRQ, COXFA4, MC4DN21, MISTR1, MLRQ...  protein-coding   \n",
      "4       [CI-9k, CI-MLRQ, COXFA4, MC4DN21, MISTR1, MLRQ...  protein-coding   \n",
      "...                                                   ...             ...   \n",
      "621501                       [NaG, Nav2.1, Nav2.2, SCN6A]  protein-coding   \n",
      "621502                       [NaG, Nav2.1, Nav2.2, SCN6A]  protein-coding   \n",
      "621503                       [NaG, Nav2.1, Nav2.2, SCN6A]  protein-coding   \n",
      "621504                       [NaG, Nav2.1, Nav2.2, SCN6A]  protein-coding   \n",
      "621505                       [NaG, Nav2.1, Nav2.2, SCN6A]  protein-coding   \n",
      "\n",
      "       map_location                                        genomic_pos  \\\n",
      "0            7p21.3  {'chr': '7', 'end': 10940153, 'ensemblgene': '...   \n",
      "1            7p21.3  {'chr': '7', 'end': 10940153, 'ensemblgene': '...   \n",
      "2            7p21.3  {'chr': '7', 'end': 10940153, 'ensemblgene': '...   \n",
      "3            7p21.3  {'chr': '7', 'end': 10940153, 'ensemblgene': '...   \n",
      "4            7p21.3  {'chr': '7', 'end': 10940153, 'ensemblgene': '...   \n",
      "...             ...                                                ...   \n",
      "621501       2q24.3  {'chr': '2', 'end': 166611482, 'ensemblgene': ...   \n",
      "621502       2q24.3  {'chr': '2', 'end': 166611482, 'ensemblgene': ...   \n",
      "621503       2q24.3  {'chr': '2', 'end': 166611482, 'ensemblgene': ...   \n",
      "621504       2q24.3  {'chr': '2', 'end': 166611482, 'ensemblgene': ...   \n",
      "621505       2q24.3  {'chr': '2', 'end': 166611482, 'ensemblgene': ...   \n",
      "\n",
      "        entrezgene   HGNC                                            summary  \n",
      "0             4697   7687  The protein encoded by this gene belongs to th...  \n",
      "1             4697   7687  The protein encoded by this gene belongs to th...  \n",
      "2             4697   7687  The protein encoded by this gene belongs to th...  \n",
      "3             4697   7687  The protein encoded by this gene belongs to th...  \n",
      "4             4697   7687  The protein encoded by this gene belongs to th...  \n",
      "...            ...    ...                                                ...  \n",
      "621501        6332  10594  This gene encodes one of the many voltage-gate...  \n",
      "621502        6332  10594  This gene encodes one of the many voltage-gate...  \n",
      "621503        6332  10594  This gene encodes one of the many voltage-gate...  \n",
      "621504        6332  10594  This gene encodes one of the many voltage-gate...  \n",
      "621505        6332  10594  This gene encodes one of the many voltage-gate...  \n",
      "\n",
      "[621506 rows x 34 columns]\n"
     ]
    }
   ],
   "source": [
    "# Merge the original pandas DataFrame (filtered_results_pd) with the gene information DataFrame on 'EnsemblID'\n",
    "filtered_results_annotated = filtered_results_pd.merge(gene_info_df, left_on='EnsemblID', right_on='query', how='left')\n",
    "\n",
    "# Drop the 'query' column (since it's the same as 'EnsemblID')\n",
    "filtered_results_annotated.drop(columns=['query'], inplace=True)\n",
    "\n",
    "# Display the annotated DataFrame\n",
    "print(filtered_results_annotated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows in filtered_results_annotated_pd: 621506\n"
     ]
    }
   ],
   "source": [
    "# Count the number of rows in filtered_results_annotated_pd\n",
    "num_rows = filtered_results_annotated_pd.shape[0]\n",
    "\n",
    "# Print the number of rows\n",
    "print(f\"Number of rows in filtered_results_annotated_pd: {num_rows}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['EnsemblID', 'VariantID', 'tss_distance', 'CAF_eQTL', 'ma_samples',\n",
      "       'ma_count', 'pval_nominal', 'Beta', 'SE', 'chromosome', 'position',\n",
      "       'OtherAlleleA', 'CodedAlleleB', 'CAF', 'AltID', 'Source',\n",
      "       'AverageMaximumPosteriorCall', 'Info', 'AA_N', 'AB_N', 'BB_N', 'TotalN',\n",
      "       'MAF', 'MissingDataProportion', 'HWE_P', 'symbol', 'name', 'alias',\n",
      "       'type_of_gene', 'map_location', 'genomic_pos', 'entrezgene', 'HGNC',\n",
      "       'summary'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# Display the headers of the DataFrame\n",
    "print(filtered_results_annotated_pd.columns)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Saving\n",
    " Let's save the annotated table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the order of columns if needed\n",
    "columns_order = ['EnsemblID', 'symbol', \n",
    "'VariantID', 'tss_distance', 'CAF_eQTL', 'ma_samples',\n",
    "'ma_count', 'pval_nominal', 'Beta', 'SE', 'chromosome', 'position',\n",
    "'OtherAlleleA', 'CodedAlleleB', 'CAF', 'AltID', 'Source',\n",
    "'AverageMaximumPosteriorCall', 'Info', 'AA_N', 'AB_N', 'BB_N', 'TotalN',\n",
    "'MAF', 'MissingDataProportion', 'HWE_P', \n",
    "'name', 'alias', 'type_of_gene', 'map_location', 'genomic_pos', 'entrezgene', 'HGNC', 'summary']\n",
    "# Ensure these columns exist in your DataFrame\n",
    "\n",
    "# Save the DataFrame with specified columns\n",
    "filtered_results_annotated_pd.to_csv(\n",
    "    os.path.join(\n",
    "        molQTL_loc, 'target_regions.nom_cis_eqtl_annotated.txt.gz'),\n",
    "    sep='\\t',\n",
    "    index=False,\n",
    "    compression='gzip',\n",
    "    columns=columns_order\n",
    ")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gwaslab",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
