{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extract 8 regions\n",
    "\n",
    "We're going to extract the 8 regions of interest (ROIs) from the _cis-eQTL_ data.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load packages\n",
    "\n",
    "Here we load the necessary packages.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to check for installation of required packages\n",
    "def check_install_package(package_name):\n",
    "    try:\n",
    "        importlib.import_module(package_name)\n",
    "    except ImportError:\n",
    "        print(f'{package_name} is not installed. Install and try again...')\n",
    "        # subprocess.check_call(['pip', 'install', package_name])\n",
    "\n",
    "import os\n",
    "import glob\n",
    "import importlib\n",
    "import subprocess\n",
    "import sys\n",
    "\n",
    "# get date and time\n",
    "from datetime import datetime\n",
    "\n",
    "# Pandas is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool\n",
    "# check_install_package('pandas')\n",
    "import pandas as pd\n",
    "\n",
    "# pyarrow is supperior to loading parquet files\n",
    "# check_install_package('pyarrow')\n",
    "import pyarrow as pa\n",
    "import pyarrow.parquet as pq\n",
    "\n",
    "# polars is a fast dataframe library\n",
    "# check_install_package('polars')\n",
    "import polars as pl\n",
    "\n",
    "# for statistical analysis\n",
    "# check_install_package('scipy')\n",
    "from scipy import stats\n",
    "import numpy as np\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also want some variables set.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking contents of the reference directory:\n",
      "\u001b[34m1000G\u001b[m\u001b[m\n",
      "\u001b[34mGoNL\u001b[m\u001b[m\n",
      "\u001b[34mHRC_r1_1_2016\u001b[m\u001b[m\n",
      "\u001b[34mHRCr11_1000Gp3v5\u001b[m\u001b[m\n",
      "\u001b[34mdbSNP\u001b[m\u001b[m\n",
      "\u001b[34mfasta\u001b[m\u001b[m\n",
      "\u001b[34mrefgenie_genomes\u001b[m\u001b[m\n",
      "\u001b[34mtcga\u001b[m\u001b[m\n",
      "\n",
      "Checking contents of the Google Drive directory:\n",
      "\u001b[34mCAC\u001b[m\u001b[m\n",
      "\u001b[34mCIMT\u001b[m\u001b[m\n",
      "\u001b[34mGIGASTROKE\u001b[m\u001b[m\n",
      "\u001b[34mMILLIONHEARTS\u001b[m\u001b[m\n",
      "\u001b[34mPGC\u001b[m\u001b[m\n",
      "\n",
      "Checking contents of the nominal cis-eQTL data directory:\n",
      "README.md\n",
      "tensorqtl_cis_nominal_chr01.cis_qtl_pairs.chr01.parquet\n",
      "tensorqtl_cis_nominal_chr02.cis_qtl_pairs.chr02.parquet\n",
      "tensorqtl_cis_nominal_chr03.cis_qtl_pairs.chr03.parquet\n",
      "tensorqtl_cis_nominal_chr04.cis_qtl_pairs.chr04.parquet\n",
      "tensorqtl_cis_nominal_chr05.cis_qtl_pairs.chr05.parquet\n",
      "tensorqtl_cis_nominal_chr06.cis_qtl_pairs.chr06.parquet\n",
      "tensorqtl_cis_nominal_chr07.cis_qtl_pairs.chr07.parquet\n",
      "tensorqtl_cis_nominal_chr08.cis_qtl_pairs.chr08.parquet\n",
      "tensorqtl_cis_nominal_chr09.cis_qtl_pairs.chr09.parquet\n",
      "tensorqtl_cis_nominal_chr10.cis_qtl_pairs.chr10.parquet\n",
      "tensorqtl_cis_nominal_chr11.cis_qtl_pairs.chr11.parquet\n",
      "tensorqtl_cis_nominal_chr12.cis_qtl_pairs.chr12.parquet\n",
      "tensorqtl_cis_nominal_chr13.cis_qtl_pairs.chr13.parquet\n",
      "tensorqtl_cis_nominal_chr14.cis_qtl_pairs.chr14.parquet\n",
      "tensorqtl_cis_nominal_chr15.cis_qtl_pairs.chr15.parquet\n",
      "tensorqtl_cis_nominal_chr16.cis_qtl_pairs.chr16.parquet\n",
      "tensorqtl_cis_nominal_chr17.cis_qtl_pairs.chr17.parquet\n",
      "tensorqtl_cis_nominal_chr18.cis_qtl_pairs.chr18.parquet\n",
      "tensorqtl_cis_nominal_chr19.cis_qtl_pairs.chr19.parquet\n",
      "tensorqtl_cis_nominal_chr20.cis_qtl_pairs.chr20.parquet\n",
      "tensorqtl_cis_nominal_chr21.cis_qtl_pairs.chr21.parquet\n",
      "tensorqtl_cis_nominal_chr22.cis_qtl_pairs.chr22.parquet\n",
      "tensorqtl_nominal_cis_qtl_pairs.annot.parquet\n",
      "tensorqtl_nominal_cis_qtl_pairs.sumstats.gwaslab\n",
      "tensorqtl_nominal_cis_qtl_pairs.sumstats.log\n",
      "tensorqtl_nominal_cis_qtl_pairs.sumstats.parquet\n",
      "tensorqtl_nominal_cis_qtl_pairs.sumstats.top.csv\n",
      "tensorqtl_nominal_cis_qtl_pairs.sumstats.top.xlsx\n",
      "\n",
      "Checking contents of the permuted cis-mQTL data directory:\n",
      "README.md\n",
      "tensormqtl.perm_cis_mqtl.txt\n",
      "tensormqtl_cis_chr1.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr10.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr11.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr12.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr13.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr14.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr15.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr16.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr17.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr18.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr19.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr2.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr20.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr21.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr22.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr3.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr4.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr5.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr6.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr7.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr8.cis_qtl.txt.gz\n",
      "tensormqtl_cis_chr9.cis_qtl.txt.gz\n",
      "\n",
      "Checking contents of the permuted trans-eQTL data directory:\n",
      "README.md\n",
      "tensorqtl_trans_full.trans_qtl_pairs.parquet\n",
      "\n",
      "Checking contents of the permuted trans-eQTL data directory:\n",
      "README.md\n",
      "tensormqtl_perm_trans_qtl_pairs.annot.parquet\n",
      "tensormqtl_trans_chr1.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr10.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr11.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr12.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr13.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr14.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr15.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr16.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr17.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr18.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr19.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr2.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr20.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr21.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr22.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr3.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr4.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr5.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr6.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr7.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr8.trans_qtl_pairs.parquet\n",
      "tensormqtl_trans_chr9.trans_qtl_pairs.parquet\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Create directories for the GWAS data and the reference data\n",
    "from subprocess import check_output\n",
    "\n",
    "# set some general defaults\n",
    "PGC_HITS = \"PGC\"\n",
    "\n",
    "POPULATION = \"EUR\"\n",
    "\n",
    "# general plotting directory\n",
    "PLOTS_loc = \"PLOTS\"\n",
    "\n",
    "# location to put molQTL results\n",
    "molQTL_loc = \"molQTL_results\"\n",
    "\n",
    "# Check if the directory exists\n",
    "if not os.path.exists(molQTL_loc):\n",
    "    # If it doesn't exist, create it\n",
    "    os.makedirs(molQTL_loc)\n",
    "\n",
    "# Check if the directory exists\n",
    "if not os.path.exists(PLOTS_loc):\n",
    "    # If it doesn't exist, create it\n",
    "    os.makedirs(PLOTS_loc)\n",
    "\n",
    "# # regional association plots directory\n",
    "# REG_PLOTS_loc = PLOTS_loc + \"/Regional_Association_Plots\"\n",
    "\n",
    "# # Check if the directory exists\n",
    "# if not os.path.exists(REG_PLOTS_loc):\n",
    "#     # If it doesn't exist, create it\n",
    "#     os.makedirs(REG_PLOTS_loc)\n",
    "\n",
    "# Reference data directory\n",
    "REF_loc = \"/Users/slaan3/PLINK/references\"\n",
    "print(\"Checking contents of the reference directory:\")\n",
    "print(check_output([\"ls\", os.path.join(REF_loc)]).decode(\"utf8\"))\n",
    "\n",
    "# GWAS data directory\n",
    "GD_loc = \"/Users/slaan3/Library/CloudStorage/GoogleDrive-s.w.vanderlaan@gmail.com/My Drive/Genomics/#Projects/TO_AITION/MR CVD MDD/GWAS\"\n",
    "print(\"Checking contents of the Google Drive directory:\")\n",
    "print(check_output([\"ls\", os.path.join(GD_loc)]).decode(\"utf8\"))\n",
    "\n",
    "# molQTL data directory\n",
    "NOM_CIS_EQTL_loc = \"/Users/slaan3/git/CirculatoryHealth/molqtl/results/version1_aernas1_firstrun/nom_cis_eqtl\"\n",
    "print(\"Checking contents of the nominal cis-eQTL data directory:\")\n",
    "print(check_output([\"ls\", os.path.join(NOM_CIS_EQTL_loc)]).decode(\"utf8\"))\n",
    "\n",
    "PERM_CIS_MQTL_loc = \"/Users/slaan3/git/CirculatoryHealth/molqtl/results/perm_cis_mqtl\"\n",
    "print(\"Checking contents of the permuted cis-mQTL data directory:\")\n",
    "print(check_output([\"ls\", os.path.join(PERM_CIS_MQTL_loc)]).decode(\"utf8\"))\n",
    "\n",
    "PERM_TRANS_EQTL_loc = \"/Users/slaan3/git/CirculatoryHealth/molqtl/results/version1_aernas1_firstrun/perm_trans_eqtl\"\n",
    "print(\"Checking contents of the permuted trans-eQTL data directory:\")\n",
    "print(check_output([\"ls\", os.path.join(PERM_TRANS_EQTL_loc)]).decode(\"utf8\"))\n",
    "\n",
    "PERM_TRANS_MQTL_loc = (\n",
    "    \"/Users/slaan3/git/CirculatoryHealth/molqtl/results/perm_trans_mqtl\"\n",
    ")\n",
    "print(\"Checking contents of the permuted trans-eQTL data directory:\")\n",
    "print(check_output([\"ls\", os.path.join(PERM_TRANS_MQTL_loc)]).decode(\"utf8\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data\n",
    "\n",
    "We are going to need the list of ROIs and the annotated cis-eQTL data.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Regions-of-interest\n",
    "\n",
    "Let's load the 8 regions we are interested in.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# polars.read_excel(\n",
    "# source: str | BytesIO | Path | BinaryIO | bytes,\n",
    "# *,\n",
    "# sheet_id: None = None,\n",
    "# sheet_name: str,\n",
    "# engine: Literal['xlsx2csv', 'openpyxl', 'pyxlsb'] | None = None,\n",
    "# xlsx2csv_options: dict[str, Any] | None = None,\n",
    "# read_csv_options: dict[str, Any] | None = None,\n",
    "# schema_overrides: SchemaDict | None = None,\n",
    "# raise_if_empty: bool = True,\n",
    "# )\n",
    "\n",
    "target_regions = pl.read_csv(\n",
    "    source=os.path.join(\"targets/regions.csv\")\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (8, 4)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>leadSNP</th><th>chromosome</th><th>lower_position</th><th>upper_position</th></tr><tr><td>str</td><td>i64</td><td>i64</td><td>i64</td></tr></thead><tbody><tr><td>&quot;7:12263538&quot;</td><td>7</td><td>11263538</td><td>13263538</td></tr><tr><td>&quot;19:4044579&quot;</td><td>19</td><td>3044579</td><td>5044579</td></tr><tr><td>&quot;15:38843887&quot;</td><td>15</td><td>37843887</td><td>39843887</td></tr><tr><td>&quot;7:114059156&quot;</td><td>7</td><td>113059156</td><td>115059156</td></tr><tr><td>&quot;7:2760750&quot;</td><td>7</td><td>1760750</td><td>3760750</td></tr><tr><td>&quot;7:1937261&quot;</td><td>7</td><td>937261</td><td>2937261</td></tr><tr><td>&quot;2:165045101&quot;</td><td>2</td><td>164045101</td><td>166045101</td></tr><tr><td>&quot;2:164915279&quot;</td><td>2</td><td>163915279</td><td>165915279</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (8, 4)\n",
       "┌─────────────┬────────────┬────────────────┬────────────────┐\n",
       "│ leadSNP     ┆ chromosome ┆ lower_position ┆ upper_position │\n",
       "│ ---         ┆ ---        ┆ ---            ┆ ---            │\n",
       "│ str         ┆ i64        ┆ i64            ┆ i64            │\n",
       "╞═════════════╪════════════╪════════════════╪════════════════╡\n",
       "│ 7:12263538  ┆ 7          ┆ 11263538       ┆ 13263538       │\n",
       "│ 19:4044579  ┆ 19         ┆ 3044579        ┆ 5044579        │\n",
       "│ 15:38843887 ┆ 15         ┆ 37843887       ┆ 39843887       │\n",
       "│ 7:114059156 ┆ 7          ┆ 113059156      ┆ 115059156      │\n",
       "│ 7:2760750   ┆ 7          ┆ 1760750        ┆ 3760750        │\n",
       "│ 7:1937261   ┆ 7          ┆ 937261         ┆ 2937261        │\n",
       "│ 2:165045101 ┆ 2          ┆ 164045101      ┆ 166045101      │\n",
       "│ 2:164915279 ┆ 2          ┆ 163915279      ┆ 165915279      │\n",
       "└─────────────┴────────────┴────────────────┴────────────────┘"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_regions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### _cis_-eQTL data\n",
    "\n",
    "We previously parsed the _cis_-eQTL data into a `.parquet` file which includes annotated results. For details check `https://github.com/CirculatoryHealth/molqtl` GitHub repository.\n",
    "\n",
    "Here we'll simply load the annotated data.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in data\n",
    "# https://stackoverflow.com/questions/33813815/how-to-read-a-parquet-file-into-pandas-dataframe\n",
    "\n",
    "import polars as pl\n",
    "\n",
    "# annotated cis-eQTLs\n",
    "sumstats_nom_cis_eqtl = pl.read_parquet(\n",
    "    source=os.path.join(\n",
    "        NOM_CIS_EQTL_loc, \"tensorqtl_nominal_cis_qtl_pairs.annot.parquet\"\n",
    "    )\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (77_285_276, 25)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>EnsemblID</th><th>VariantID</th><th>tss_distance</th><th>CAF_eQTL</th><th>ma_samples</th><th>ma_count</th><th>pval_nominal</th><th>Beta</th><th>SE</th><th>chromosome</th><th>position</th><th>OtherAlleleA</th><th>CodedAlleleB</th><th>CAF</th><th>AltID</th><th>Source</th><th>AverageMaximumPosteriorCall</th><th>Info</th><th>AA_N</th><th>AB_N</th><th>BB_N</th><th>TotalN</th><th>MAF</th><th>MissingDataProportion</th><th>HWE_P</th></tr><tr><td>str</td><td>str</td><td>i32</td><td>f32</td><td>i32</td><td>i32</td><td>f64</td><td>f32</td><td>f32</td><td>i64</td><td>i64</td><td>str</td><td>str</td><td>f64</td><td>str</td><td>str</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>i64</td><td>f64</td><td>f64</td><td>f64</td></tr></thead><tbody><tr><td>&quot;ENSG00000187634&quot;</td><td>&quot;1:693731&quot;</td><td>-240519</td><td>0.134185</td><td>153</td><td>168</td><td>0.881492</td><td>0.005953</td><td>0.039917</td><td>1</td><td>693731</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.137241</td><td>&quot;rs12238997&quot;</td><td>&quot;HRCr11&quot;</td><td>0.902046</td><td>0.624501</td><td>1637.01</td><td>451.103</td><td>35.569</td><td>2124</td><td>0.122957</td><td>0.000075</td><td>0.544841</td></tr><tr><td>&quot;ENSG00000187634&quot;</td><td>&quot;1:714596&quot;</td><td>-219654</td><td>0.038339</td><td>48</td><td>48</td><td>0.834746</td><td>-0.015369</td><td>0.073635</td><td>1</td><td>714596</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.0327213</td><td>&quot;rs149887893&quot;</td><td>&quot;HRCr11&quot;</td><td>0.968278</td><td>0.577564</td><td>1987.73</td><td>135.495</td><td>0.677</td><td>2124</td><td>0.0322165</td><td>0.000024</td><td>0.272504</td></tr><tr><td>&quot;ENSG00000187634&quot;</td><td>&quot;1:715367&quot;</td><td>-218883</td><td>0.038339</td><td>48</td><td>48</td><td>0.834746</td><td>-0.015369</td><td>0.073635</td><td>1</td><td>715367</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.0324859</td><td>&quot;rs12184277&quot;</td><td>&quot;HRCr11&quot;</td><td>0.975207</td><td>0.67196</td><td>1976.85</td><td>146.512</td><td>0.556</td><td>2124</td><td>0.0347528</td><td>0.000019</td><td>0.176551</td></tr><tr><td>&quot;ENSG00000187634&quot;</td><td>&quot;1:717485&quot;</td><td>-216765</td><td>0.038339</td><td>48</td><td>48</td><td>0.834746</td><td>-0.015369</td><td>0.073635</td><td>1</td><td>717485</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>0.0324859</td><td>&quot;rs12184279&quot;</td><td>&quot;HRCr11&quot;</td><td>0.975183</td><td>0.670147</td><td>1978.35</td><td>145.041</td><td>0.517</td><td>2124</td><td>0.0343883</td><td>0.000022</td><td>0.174817</td></tr><tr><td>&quot;ENSG00000187634&quot;</td><td>&quot;1:720381&quot;</td><td>-213869</td><td>0.038339</td><td>48</td><td>48</td><td>0.834746</td><td>-0.015369</td><td>0.073635</td><td>1</td><td>720381</td><td>&quot;G&quot;</td><td>&quot;T&quot;</td><td>0.0327213</td><td>&quot;rs116801199&quot;</td><td>&quot;HRCr11&quot;</td><td>0.974476</td><td>0.667536</td><td>1973.72</td><td>149.503</td><td>0.691</td><td>2124</td><td>0.0355205</td><td>0.00002</td><td>0.110452</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;ENSG00000079974&quot;</td><td>&quot;22:51220249&quot;</td><td>444664</td><td>0.035942</td><td>43</td><td>45</td><td>0.863512</td><td>-0.004492</td><td>0.026119</td><td>22</td><td>51220249</td><td>&quot;A&quot;</td><td>&quot;T&quot;</td><td>0.0442561</td><td>&quot;rs9616977&quot;</td><td>&quot;1000Gp3v5&quot;</td><td>0.913151</td><td>0.403798</td><td>1832.6</td><td>278.562</td><td>12.692</td><td>2124</td><td>0.0715553</td><td>0.000034</td><td>0.622809</td></tr><tr><td>&quot;ENSG00000079974&quot;</td><td>&quot;22:51221190&quot;</td><td>445605</td><td>0.035144</td><td>42</td><td>44</td><td>0.671267</td><td>0.011112</td><td>0.026169</td><td>22</td><td>51221190</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>0.0400188</td><td>&quot;rs369304721&quot;</td><td>&quot;HRCr11&quot;</td><td>0.933398</td><td>0.443438</td><td>1911.83</td><td>204.73</td><td>7.373</td><td>2124</td><td>0.0516673</td><td>0.000015</td><td>0.500544</td></tr><tr><td>&quot;ENSG00000079974&quot;</td><td>&quot;22:51221731&quot;</td><td>446146</td><td>0.035942</td><td>43</td><td>45</td><td>0.863512</td><td>-0.004492</td><td>0.026119</td><td>22</td><td>51221731</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.0442561</td><td>&quot;rs115055839&quot;</td><td>&quot;HRCr11&quot;</td><td>0.911894</td><td>0.405944</td><td>1826.36</td><td>284.126</td><td>13.448</td><td>2124</td><td>0.0732185</td><td>0.000016</td><td>0.523538</td></tr><tr><td>&quot;ENSG00000079974&quot;</td><td>&quot;22:51222100&quot;</td><td>446515</td><td>0.054313</td><td>67</td><td>68</td><td>0.408353</td><td>0.018264</td><td>0.022074</td><td>22</td><td>51222100</td><td>&quot;G&quot;</td><td>&quot;T&quot;</td><td>0.0503766</td><td>&quot;rs114553188&quot;</td><td>&quot;HRCr11&quot;</td><td>0.969267</td><td>0.716716</td><td>1911.95</td><td>205.392</td><td>6.52</td><td>2124</td><td>0.0514234</td><td>0.000033</td><td>0.821168</td></tr><tr><td>&quot;ENSG00000079974&quot;</td><td>&quot;22:51229805&quot;</td><td>454220</td><td>0.038339</td><td>46</td><td>48</td><td>0.973919</td><td>-0.000827</td><td>0.025284</td><td>22</td><td>51229805</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.0477872</td><td>&quot;rs9616985&quot;</td><td>&quot;HRCr11&quot;</td><td>0.935419</td><td>0.585</td><td>1821.25</td><td>287.583</td><td>15.116</td><td>2124</td><td>0.0748171</td><td>0.000012</td><td>0.342973</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (77_285_276, 25)\n",
       "┌────────────┬────────────┬────────────┬──────────┬───┬────────┬───────────┬────────────┬──────────┐\n",
       "│ EnsemblID  ┆ VariantID  ┆ tss_distan ┆ CAF_eQTL ┆ … ┆ TotalN ┆ MAF       ┆ MissingDat ┆ HWE_P    │\n",
       "│ ---        ┆ ---        ┆ ce         ┆ ---      ┆   ┆ ---    ┆ ---       ┆ aProportio ┆ ---      │\n",
       "│ str        ┆ str        ┆ ---        ┆ f32      ┆   ┆ i64    ┆ f64       ┆ n          ┆ f64      │\n",
       "│            ┆            ┆ i32        ┆          ┆   ┆        ┆           ┆ ---        ┆          │\n",
       "│            ┆            ┆            ┆          ┆   ┆        ┆           ┆ f64        ┆          │\n",
       "╞════════════╪════════════╪════════════╪══════════╪═══╪════════╪═══════════╪════════════╪══════════╡\n",
       "│ ENSG000001 ┆ 1:693731   ┆ -240519    ┆ 0.134185 ┆ … ┆ 2124   ┆ 0.122957  ┆ 0.000075   ┆ 0.544841 │\n",
       "│ 87634      ┆            ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
       "│ ENSG000001 ┆ 1:714596   ┆ -219654    ┆ 0.038339 ┆ … ┆ 2124   ┆ 0.0322165 ┆ 0.000024   ┆ 0.272504 │\n",
       "│ 87634      ┆            ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
       "│ ENSG000001 ┆ 1:715367   ┆ -218883    ┆ 0.038339 ┆ … ┆ 2124   ┆ 0.0347528 ┆ 0.000019   ┆ 0.176551 │\n",
       "│ 87634      ┆            ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
       "│ ENSG000001 ┆ 1:717485   ┆ -216765    ┆ 0.038339 ┆ … ┆ 2124   ┆ 0.0343883 ┆ 0.000022   ┆ 0.174817 │\n",
       "│ 87634      ┆            ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
       "│ ENSG000001 ┆ 1:720381   ┆ -213869    ┆ 0.038339 ┆ … ┆ 2124   ┆ 0.0355205 ┆ 0.00002    ┆ 0.110452 │\n",
       "│ 87634      ┆            ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
       "│ …          ┆ …          ┆ …          ┆ …        ┆ … ┆ …      ┆ …         ┆ …          ┆ …        │\n",
       "│ ENSG000000 ┆ 22:5122024 ┆ 444664     ┆ 0.035942 ┆ … ┆ 2124   ┆ 0.0715553 ┆ 0.000034   ┆ 0.622809 │\n",
       "│ 79974      ┆ 9          ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
       "│ ENSG000000 ┆ 22:5122119 ┆ 445605     ┆ 0.035144 ┆ … ┆ 2124   ┆ 0.0516673 ┆ 0.000015   ┆ 0.500544 │\n",
       "│ 79974      ┆ 0          ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
       "│ ENSG000000 ┆ 22:5122173 ┆ 446146     ┆ 0.035942 ┆ … ┆ 2124   ┆ 0.0732185 ┆ 0.000016   ┆ 0.523538 │\n",
       "│ 79974      ┆ 1          ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
       "│ ENSG000000 ┆ 22:5122210 ┆ 446515     ┆ 0.054313 ┆ … ┆ 2124   ┆ 0.0514234 ┆ 0.000033   ┆ 0.821168 │\n",
       "│ 79974      ┆ 0          ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
       "│ ENSG000000 ┆ 22:5122980 ┆ 454220     ┆ 0.038339 ┆ … ┆ 2124   ┆ 0.0748171 ┆ 0.000012   ┆ 0.342973 │\n",
       "│ 79974      ┆ 5          ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
       "└────────────┴────────────┴────────────┴──────────┴───┴────────┴───────────┴────────────┴──────────┘"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sumstats_nom_cis_eqtl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### _cis_-mQTL data\n",
    "\n",
    "We previously parsed the _cis_-mQTL data into a `.parquet` file which includes annotated results. For details check `https://github.com/CirculatoryHealth/molqtl` GitHub repository.\n",
    "\n",
    "Here we'll simply load the annotated data.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in data\n",
    "# https://stackoverflow.com/questions/33813815/how-to-read-a-parquet-file-into-pandas-dataframe\n",
    "\n",
    "import polars as pl\n",
    "\n",
    "# Define the schema for your columns, e.g., `num_var` as an integer (or other types if necessary)\n",
    "schema_overrides = {\n",
    "    \"num_var\": pl.Int64  # or pl.Float64 if you expect floats, or pl.Utf8 if it's text data\n",
    "}\n",
    "# annotated cis-mQTLs\n",
    "sumstats_nom_cis_mqtl = pl.read_csv(\n",
    "    source=os.path.join(PERM_CIS_MQTL_loc, \"tensormqtl.perm_cis_mqtl.txt\"),\n",
    "    separator='\\t',\n",
    "    schema_overrides=schema_overrides,\n",
    "    infer_schema_length=10000,  # Increase if needed\n",
    "    ignore_errors=True  # Skips rows with parsing errors\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (268_375, 18)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>phenotype_id</th><th>num_var</th><th>beta_shape1</th><th>beta_shape2</th><th>true_df</th><th>pval_true_df</th><th>variant_id</th><th>tss_distance</th><th>ma_samples</th><th>ma_count</th><th>af</th><th>pval_nominal</th><th>slope</th><th>slope_se</th><th>pval_perm</th><th>pval_beta</th><th>qval</th><th>pval_nominal_threshold</th></tr><tr><td>str</td><td>i64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>str</td><td>i64</td><td>i64</td><td>i64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td></tr></thead><tbody><tr><td>&quot;cg21870274&quot;</td><td>309</td><td>1.01037</td><td>15.0882</td><td>351.533</td><td>0.0495503</td><td>&quot;1:752307&quot;</td><td>682715</td><td>15</td><td>15</td><td>0.0170068</td><td>0.036876</td><td>-0.406806</td><td>0.194254</td><td>0.531247</td><td>0.530703</td><td>0.468602</td><td>0.001232</td></tr><tr><td>&quot;cg08258224&quot;</td><td>731</td><td>1.03658</td><td>40.9219</td><td>342.718</td><td>0.101636</td><td>&quot;1:1636400&quot;</td><td>836316</td><td>37</td><td>38</td><td>0.0430839</td><td>0.0780652</td><td>-0.160865</td><td>0.0910595</td><td>0.984102</td><td>0.986523</td><td>0.605861</td><td>0.000509</td></tr><tr><td>&quot;cg18147296&quot;</td><td>750</td><td>1.02594</td><td>43.8044</td><td>350.307</td><td>0.000046</td><td>&quot;1:771410&quot;</td><td>-41130</td><td>130</td><td>141</td><td>0.840136</td><td>0.000014</td><td>-0.195278</td><td>0.0444456</td><td>0.0017</td><td>0.001693</td><td>0.006203</td><td>0.000454</td></tr><tr><td>&quot;cg13938959&quot;</td><td>787</td><td>1.00503</td><td>41.991</td><td>346.635</td><td>0.0160055</td><td>&quot;1:800193&quot;</td><td>-33991</td><td>7</td><td>7</td><td>0.007937</td><td>0.009935</td><td>0.568674</td><td>0.219518</td><td>0.49525</td><td>0.48971</td><td>0.45302</td><td>0.000433</td></tr><tr><td>&quot;cg12445832&quot;</td><td>787</td><td>1.01013</td><td>42.2714</td><td>344.888</td><td>0.003059</td><td>&quot;1:1706886&quot;</td><td>872590</td><td>44</td><td>46</td><td>0.0521542</td><td>0.001483</td><td>-0.25942</td><td>0.08106</td><td>0.118388</td><td>0.118428</td><td>0.211842</td><td>0.000439</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;cg19004771&quot;</td><td>2979</td><td>1.04971</td><td>202.825</td><td>323.792</td><td>0.000565</td><td>&quot;22:51181759&quot;</td><td>8693</td><td>100</td><td>108</td><td>0.877551</td><td>0.000134</td><td>-0.238654</td><td>0.0618922</td><td>0.0933907</td><td>0.0950843</td><td>0.152963</td><td>0.000135</td></tr><tr><td>&quot;cg20569369&quot;</td><td>2979</td><td>1.04484</td><td>216.682</td><td>329.092</td><td>0.00114</td><td>&quot;22:50519640&quot;</td><td>-654007</td><td>25</td><td>25</td><td>0.0283447</td><td>0.000352</td><td>-0.456097</td><td>0.126513</td><td>0.20348</td><td>0.201155</td><td>0.249547</td><td>0.000124</td></tr><tr><td>&quot;cg26034629&quot;</td><td>2975</td><td>1.03817</td><td>214.838</td><td>329.486</td><td>0.003815</td><td>&quot;22:50913182&quot;</td><td>-262183</td><td>75</td><td>79</td><td>0.0895692</td><td>0.001493</td><td>0.316939</td><td>0.0990935</td><td>0.549345</td><td>0.542439</td><td>0.423997</td><td>0.000122</td></tr><tr><td>&quot;cg25232725&quot;</td><td>2975</td><td>1.04781</td><td>217.38</td><td>327.78</td><td>0.001621</td><td>&quot;22:50642777&quot;</td><td>-532739</td><td>302</td><td>394</td><td>0.553288</td><td>0.000522</td><td>-0.164639</td><td>0.0470652</td><td>0.278472</td><td>0.275782</td><td>0.300416</td><td>0.000125</td></tr><tr><td>&quot;cg05615487&quot;</td><td>2973</td><td>1.05074</td><td>219.59</td><td>329.205</td><td>0.004493</td><td>&quot;22:51151350&quot;</td><td>-25058</td><td>270</td><td>335</td><td>0.379819</td><td>0.001805</td><td>0.161365</td><td>0.0513606</td><td>0.610839</td><td>0.60608</td><td>0.445053</td><td>0.000126</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (268_375, 18)\n",
       "┌────────────┬─────────┬────────────┬───────────┬───┬───────────┬───────────┬──────────┬───────────┐\n",
       "│ phenotype_ ┆ num_var ┆ beta_shape ┆ beta_shap ┆ … ┆ pval_perm ┆ pval_beta ┆ qval     ┆ pval_nomi │\n",
       "│ id         ┆ ---     ┆ 1          ┆ e2        ┆   ┆ ---       ┆ ---       ┆ ---      ┆ nal_thres │\n",
       "│ ---        ┆ i64     ┆ ---        ┆ ---       ┆   ┆ f64       ┆ f64       ┆ f64      ┆ hold      │\n",
       "│ str        ┆         ┆ f64        ┆ f64       ┆   ┆           ┆           ┆          ┆ ---       │\n",
       "│            ┆         ┆            ┆           ┆   ┆           ┆           ┆          ┆ f64       │\n",
       "╞════════════╪═════════╪════════════╪═══════════╪═══╪═══════════╪═══════════╪══════════╪═══════════╡\n",
       "│ cg21870274 ┆ 309     ┆ 1.01037    ┆ 15.0882   ┆ … ┆ 0.531247  ┆ 0.530703  ┆ 0.468602 ┆ 0.001232  │\n",
       "│ cg08258224 ┆ 731     ┆ 1.03658    ┆ 40.9219   ┆ … ┆ 0.984102  ┆ 0.986523  ┆ 0.605861 ┆ 0.000509  │\n",
       "│ cg18147296 ┆ 750     ┆ 1.02594    ┆ 43.8044   ┆ … ┆ 0.0017    ┆ 0.001693  ┆ 0.006203 ┆ 0.000454  │\n",
       "│ cg13938959 ┆ 787     ┆ 1.00503    ┆ 41.991    ┆ … ┆ 0.49525   ┆ 0.48971   ┆ 0.45302  ┆ 0.000433  │\n",
       "│ cg12445832 ┆ 787     ┆ 1.01013    ┆ 42.2714   ┆ … ┆ 0.118388  ┆ 0.118428  ┆ 0.211842 ┆ 0.000439  │\n",
       "│ …          ┆ …       ┆ …          ┆ …         ┆ … ┆ …         ┆ …         ┆ …        ┆ …         │\n",
       "│ cg19004771 ┆ 2979    ┆ 1.04971    ┆ 202.825   ┆ … ┆ 0.0933907 ┆ 0.0950843 ┆ 0.152963 ┆ 0.000135  │\n",
       "│ cg20569369 ┆ 2979    ┆ 1.04484    ┆ 216.682   ┆ … ┆ 0.20348   ┆ 0.201155  ┆ 0.249547 ┆ 0.000124  │\n",
       "│ cg26034629 ┆ 2975    ┆ 1.03817    ┆ 214.838   ┆ … ┆ 0.549345  ┆ 0.542439  ┆ 0.423997 ┆ 0.000122  │\n",
       "│ cg25232725 ┆ 2975    ┆ 1.04781    ┆ 217.38    ┆ … ┆ 0.278472  ┆ 0.275782  ┆ 0.300416 ┆ 0.000125  │\n",
       "│ cg05615487 ┆ 2973    ┆ 1.05074    ┆ 219.59    ┆ … ┆ 0.610839  ┆ 0.60608   ┆ 0.445053 ┆ 0.000126  │\n",
       "└────────────┴─────────┴────────────┴───────────┴───┴───────────┴───────────┴──────────┴───────────┘"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sumstats_nom_cis_mqtl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract _cis_-eQTL data\n",
    "\n",
    "Now we're ready to extract the data for each region in one table.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape: (621_506, 25)\n",
      "┌────────────┬────────────┬────────────┬──────────┬───┬────────┬───────────┬────────────┬──────────┐\n",
      "│ EnsemblID  ┆ VariantID  ┆ tss_distan ┆ CAF_eQTL ┆ … ┆ TotalN ┆ MAF       ┆ MissingDat ┆ HWE_P    │\n",
      "│ ---        ┆ ---        ┆ ce         ┆ ---      ┆   ┆ ---    ┆ ---       ┆ aProportio ┆ ---      │\n",
      "│ str        ┆ str        ┆ ---        ┆ f32      ┆   ┆ i64    ┆ f64       ┆ n          ┆ f64      │\n",
      "│            ┆            ┆ i32        ┆          ┆   ┆        ┆           ┆ ---        ┆          │\n",
      "│            ┆            ┆            ┆          ┆   ┆        ┆           ┆ f64        ┆          │\n",
      "╞════════════╪════════════╪════════════╪══════════╪═══╪════════╪═══════════╪════════════╪══════════╡\n",
      "│ ENSG000001 ┆ 7:11263538 ┆ 327489     ┆ 0.071086 ┆ … ┆ 2124   ┆ 0.0684027 ┆ 0.000001   ┆ 0.610073 │\n",
      "│ 89043      ┆            ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
      "│ ENSG000001 ┆ 7:11264536 ┆ 328487     ┆ 0.047923 ┆ … ┆ 2124   ┆ 0.0536334 ┆ 0.000004   ┆ 0.830364 │\n",
      "│ 89043      ┆            ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
      "│ ENSG000001 ┆ 7:11264670 ┆ 328621     ┆ 0.075879 ┆ … ┆ 2124   ┆ 0.0805431 ┆ 0.000006   ┆ 0.304873 │\n",
      "│ 89043      ┆            ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
      "│ ENSG000001 ┆ 7:11264692 ┆ 328643     ┆ 0.521565 ┆ … ┆ 2124   ┆ 0.45498   ┆ 0.000005   ┆ 1.0      │\n",
      "│ 89043      ┆            ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
      "│ ENSG000001 ┆ 7:11265421 ┆ 329372     ┆ 0.598243 ┆ … ┆ 2124   ┆ 0.38279   ┆ 0.000004   ┆ 0.408573 │\n",
      "│ 89043      ┆            ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
      "│ …          ┆ …          ┆ …          ┆ …        ┆ … ┆ …      ┆ …         ┆ …          ┆ …        │\n",
      "│ ENSG000001 ┆ 2:16591356 ┆ -593964    ┆ 0.646965 ┆ … ┆ 2124   ┆ 0.350612  ┆ 0.000003   ┆ 0.390789 │\n",
      "│ 36546      ┆ 4          ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
      "│ ENSG000001 ┆ 2:16591424 ┆ -593287    ┆ 0.162939 ┆ … ┆ 2124   ┆ 0.17201   ┆ 0.000009   ┆ 0.322517 │\n",
      "│ 36546      ┆ 1          ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
      "│ ENSG000001 ┆ 2:16591435 ┆ -593175    ┆ 0.227636 ┆ … ┆ 2124   ┆ 0.226552  ┆ 0.000008   ┆ 0.292141 │\n",
      "│ 36546      ┆ 3          ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
      "│ ENSG000001 ┆ 2:16591441 ┆ -593112    ┆ 0.234026 ┆ … ┆ 2124   ┆ 0.234516  ┆ 0.000008   ┆ 0.671834 │\n",
      "│ 36546      ┆ 6          ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
      "│ ENSG000001 ┆ 2:16591480 ┆ -592726    ┆ 0.13738  ┆ … ┆ 2124   ┆ 0.137259  ┆ 0.000004   ┆ 0.141703 │\n",
      "│ 36546      ┆ 2          ┆            ┆          ┆   ┆        ┆           ┆            ┆          │\n",
      "└────────────┴────────────┴────────────┴──────────┴───┴────────┴───────────┴────────────┴──────────┘\n"
     ]
    }
   ],
   "source": [
    "import polars as pl\n",
    "\n",
    "# Initialize an empty DataFrame with the same schema as sumstats_nom_cis_eqtl\n",
    "filtered_results = pl.DataFrame(schema=sumstats_nom_cis_eqtl.schema)\n",
    "\n",
    "# Loop through each row of df2 and filter df1 based on chromosome and position range\n",
    "for row in target_regions.iter_rows(named=True):\n",
    "    chrom = row['chromosome']\n",
    "    lower_pos = row['lower_position']\n",
    "    upper_pos = row['upper_position']\n",
    "    \n",
    "    # Filter df1 based on the chromosome and position range using polars\n",
    "    filtered = sumstats_nom_cis_eqtl.filter(\n",
    "        (sumstats_nom_cis_eqtl['chromosome'] == chrom) & \n",
    "        (sumstats_nom_cis_eqtl['position'] >= lower_pos) & \n",
    "        (sumstats_nom_cis_eqtl['position'] <= upper_pos)\n",
    "    )\n",
    "    \n",
    "    # Append the filtered rows to the result DataFrame\n",
    "    filtered_results = filtered_results.vstack(filtered)\n",
    "\n",
    "# Show the final filtered DataFrame\n",
    "print(filtered_results)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows in filtered_results: 621506\n"
     ]
    }
   ],
   "source": [
    "# Count the number of rows in the filtered_results DataFrame\n",
    "num_rows = filtered_results.shape[0]\n",
    "\n",
    "# Print the number of rows\n",
    "print(f\"Number of rows in filtered_results: {num_rows}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "del sumstats_nom_cis_eqtl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Annotation\n",
    "\n",
    "Let's also annotate this table with more information.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>alias</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>map_location</th>\n",
       "      <th>genomic_pos</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>HGNC</th>\n",
       "      <th>summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000105248</td>\n",
       "      <td>YJU2</td>\n",
       "      <td>YJU2 splicing factor homolog</td>\n",
       "      <td>CCDC94</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>19p13.3</td>\n",
       "      <td>{'chr': '19', 'end': 4269088, 'ensemblgene': '...</td>\n",
       "      <td>55702</td>\n",
       "      <td>25518</td>\n",
       "      <td>Predicted to enable metal ion binding activity...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000188549</td>\n",
       "      <td>CCDC9B</td>\n",
       "      <td>coiled-coil domain containing 9B</td>\n",
       "      <td>C15orf52</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>15q15.1</td>\n",
       "      <td>{'chr': '15', 'end': 40340939, 'ensemblgene': ...</td>\n",
       "      <td>388115</td>\n",
       "      <td>33488</td>\n",
       "      <td>Enables RNA binding activity. [provided by All...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000157778</td>\n",
       "      <td>PSMG3</td>\n",
       "      <td>proteasome assembly chaperone 3</td>\n",
       "      <td>[C7orf48, PAC3, Pba3]</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>7p22.3</td>\n",
       "      <td>{'chr': '7', 'end': 1571005, 'ensemblgene': 'E...</td>\n",
       "      <td>84262</td>\n",
       "      <td>22420</td>\n",
       "      <td>Enables molecular adaptor activity. Involved i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000146530</td>\n",
       "      <td>VWDE</td>\n",
       "      <td>von Willebrand factor D and EGF domains</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>7p21.3</td>\n",
       "      <td>{'chr': '7', 'end': 12403941, 'ensemblgene': '...</td>\n",
       "      <td>221806</td>\n",
       "      <td>21897</td>\n",
       "      <td>Predicted to enable signaling receptor binding...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000106268</td>\n",
       "      <td>NUDT1</td>\n",
       "      <td>nudix hydrolase 1</td>\n",
       "      <td>MTH1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>7p22.3</td>\n",
       "      <td>{'chr': '7', 'end': 2251146, 'ensemblgene': 'E...</td>\n",
       "      <td>4521</td>\n",
       "      <td>8048</td>\n",
       "      <td>Misincorporation of oxidized nucleoside tripho...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>ENSG00000140320</td>\n",
       "      <td>BAHD1</td>\n",
       "      <td>bromo adjacent homology domain containing 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>15q15.1</td>\n",
       "      <td>{'chr': '15', 'end': 40468236, 'ensemblgene': ...</td>\n",
       "      <td>22893</td>\n",
       "      <td>29153</td>\n",
       "      <td>Enables chromatin binding activity. Involved i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>ENSG00000172000</td>\n",
       "      <td>ZNF556</td>\n",
       "      <td>zinc finger protein 556</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>19p13.3</td>\n",
       "      <td>{'chr': '19', 'end': 2883445, 'ensemblgene': '...</td>\n",
       "      <td>80032</td>\n",
       "      <td>25669</td>\n",
       "      <td>Predicted to enable DNA-binding transcription ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>ENSG00000073067</td>\n",
       "      <td>CYP2W1</td>\n",
       "      <td>cytochrome P450 family 2 subfamily W member 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>7p22.3</td>\n",
       "      <td>{'chr': '7', 'end': 989640, 'ensemblgene': 'EN...</td>\n",
       "      <td>54905</td>\n",
       "      <td>20243</td>\n",
       "      <td>This gene encodes a member of the cytochrome P...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>ENSG00000005108</td>\n",
       "      <td>THSD7A</td>\n",
       "      <td>thrombospondin type 1 domain containing 7A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>7p21.3</td>\n",
       "      <td>{'chr': '7', 'end': 11832198, 'ensemblgene': '...</td>\n",
       "      <td>221981</td>\n",
       "      <td>22207</td>\n",
       "      <td>The protein encoded by this gene is found almo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>ENSG00000105298</td>\n",
       "      <td>CACTIN</td>\n",
       "      <td>cactin, spliceosome C complex subunit</td>\n",
       "      <td>[C19orf29, NY-REN-24, fSAPc]</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>19p13.3</td>\n",
       "      <td>{'chr': '19', 'end': 3626815, 'ensemblgene': '...</td>\n",
       "      <td>58509</td>\n",
       "      <td>29938</td>\n",
       "      <td>Enables RNA binding activity. Involved in seve...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>176 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               query  symbol                                           name  \\\n",
       "0    ENSG00000105248    YJU2                   YJU2 splicing factor homolog   \n",
       "1    ENSG00000188549  CCDC9B               coiled-coil domain containing 9B   \n",
       "2    ENSG00000157778   PSMG3                proteasome assembly chaperone 3   \n",
       "3    ENSG00000146530    VWDE        von Willebrand factor D and EGF domains   \n",
       "4    ENSG00000106268   NUDT1                              nudix hydrolase 1   \n",
       "..               ...     ...                                            ...   \n",
       "171  ENSG00000140320   BAHD1    bromo adjacent homology domain containing 1   \n",
       "172  ENSG00000172000  ZNF556                        zinc finger protein 556   \n",
       "173  ENSG00000073067  CYP2W1  cytochrome P450 family 2 subfamily W member 1   \n",
       "174  ENSG00000005108  THSD7A     thrombospondin type 1 domain containing 7A   \n",
       "175  ENSG00000105298  CACTIN          cactin, spliceosome C complex subunit   \n",
       "\n",
       "                            alias    type_of_gene map_location  \\\n",
       "0                          CCDC94  protein-coding      19p13.3   \n",
       "1                        C15orf52  protein-coding      15q15.1   \n",
       "2           [C7orf48, PAC3, Pba3]  protein-coding       7p22.3   \n",
       "3                             NaN  protein-coding       7p21.3   \n",
       "4                            MTH1  protein-coding       7p22.3   \n",
       "..                            ...             ...          ...   \n",
       "171                           NaN  protein-coding      15q15.1   \n",
       "172                           NaN  protein-coding      19p13.3   \n",
       "173                           NaN  protein-coding       7p22.3   \n",
       "174                           NaN  protein-coding       7p21.3   \n",
       "175  [C19orf29, NY-REN-24, fSAPc]  protein-coding      19p13.3   \n",
       "\n",
       "                                           genomic_pos entrezgene   HGNC  \\\n",
       "0    {'chr': '19', 'end': 4269088, 'ensemblgene': '...      55702  25518   \n",
       "1    {'chr': '15', 'end': 40340939, 'ensemblgene': ...     388115  33488   \n",
       "2    {'chr': '7', 'end': 1571005, 'ensemblgene': 'E...      84262  22420   \n",
       "3    {'chr': '7', 'end': 12403941, 'ensemblgene': '...     221806  21897   \n",
       "4    {'chr': '7', 'end': 2251146, 'ensemblgene': 'E...       4521   8048   \n",
       "..                                                 ...        ...    ...   \n",
       "171  {'chr': '15', 'end': 40468236, 'ensemblgene': ...      22893  29153   \n",
       "172  {'chr': '19', 'end': 2883445, 'ensemblgene': '...      80032  25669   \n",
       "173  {'chr': '7', 'end': 989640, 'ensemblgene': 'EN...      54905  20243   \n",
       "174  {'chr': '7', 'end': 11832198, 'ensemblgene': '...     221981  22207   \n",
       "175  {'chr': '19', 'end': 3626815, 'ensemblgene': '...      58509  29938   \n",
       "\n",
       "                                               summary  \n",
       "0    Predicted to enable metal ion binding activity...  \n",
       "1    Enables RNA binding activity. [provided by All...  \n",
       "2    Enables molecular adaptor activity. Involved i...  \n",
       "3    Predicted to enable signaling receptor binding...  \n",
       "4    Misincorporation of oxidized nucleoside tripho...  \n",
       "..                                                 ...  \n",
       "171  Enables chromatin binding activity. Involved i...  \n",
       "172  Predicted to enable DNA-binding transcription ...  \n",
       "173  This gene encodes a member of the cytochrome P...  \n",
       "174  The protein encoded by this gene is found almo...  \n",
       "175  Enables RNA binding activity. Involved in seve...  \n",
       "\n",
       "[176 rows x 10 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import mygene\n",
    "import pandas as pd\n",
    "import polars as pl\n",
    "\n",
    "# Initialize mygene client\n",
    "mg = mygene.MyGeneInfo()\n",
    "\n",
    "# Convert the polars DataFrame (filtered_results) to a pandas DataFrame\n",
    "filtered_results_pd = filtered_results.to_pandas()\n",
    "\n",
    "# Get the unique Ensembl IDs from the polars DataFrame (filtered_results)\n",
    "ensembl_ids = filtered_results['EnsemblID'].unique().to_list()\n",
    "\n",
    "# Query mygene to get additional gene information (e.g., symbol, name, etc.)\n",
    "gene_info = mg.querymany(ensembl_ids, \n",
    "                         scopes='ensembl.gene', \n",
    "                         fields='symbol,name,alias,type_of_gene,map_location,genomic_pos,entrezgene,HGNC,summary', \n",
    "                         species='human')\n",
    "\n",
    "# Convert the gene info to a pandas DataFrame\n",
    "gene_info_df = pd.DataFrame(gene_info)\n",
    "\n",
    "# Select relevant columns\n",
    "gene_info_df = gene_info_df[['query', 'symbol', 'name', 'alias', 'type_of_gene', 'map_location', 'genomic_pos', 'entrezgene', 'HGNC', 'summary']]\n",
    "\n",
    "# Convert the pandas DataFrame back to polars DataFrame\n",
    "gene_info_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              EnsemblID    VariantID  tss_distance  CAF_eQTL  ma_samples  \\\n",
      "0       ENSG00000189043   7:11263538        327489  0.071086          84   \n",
      "1       ENSG00000189043   7:11264536        328487  0.047923          56   \n",
      "2       ENSG00000189043   7:11264670        328621  0.075879          88   \n",
      "3       ENSG00000189043   7:11264692        328643  0.521565         455   \n",
      "4       ENSG00000189043   7:11265421        329372  0.598243         404   \n",
      "...                 ...          ...           ...       ...         ...   \n",
      "621501  ENSG00000136546  2:165913564       -593964  0.646965         362   \n",
      "621502  ENSG00000136546  2:165914241       -593287  0.162939         186   \n",
      "621503  ENSG00000136546  2:165914353       -593175  0.227636         251   \n",
      "621504  ENSG00000136546  2:165914416       -593112  0.234026         259   \n",
      "621505  ENSG00000136546  2:165914802       -592726  0.137380         155   \n",
      "\n",
      "        ma_count  pval_nominal      Beta        SE  chromosome  ...     HWE_P  \\\n",
      "0             89      0.337176 -0.027423  0.028549           7  ...  0.610073   \n",
      "1             60      0.349662  0.032065  0.034257           7  ...  0.830364   \n",
      "2             95      0.299686 -0.028301  0.027264           7  ...  0.304873   \n",
      "3            599      0.789388 -0.004011  0.015009           7  ...  1.000000   \n",
      "4            503      0.354785 -0.014423  0.015574           7  ...  0.408573   \n",
      "...          ...           ...       ...       ...         ...  ...       ...   \n",
      "621501       442      0.026601  0.039790  0.017899           2  ...  0.390789   \n",
      "621502       204      0.591413 -0.012557  0.023379           2  ...  0.322517   \n",
      "621503       285      0.338273  0.019565  0.020414           2  ...  0.292141   \n",
      "621504       293      0.031736  0.043972  0.020424           2  ...  0.671834   \n",
      "621505       172      0.873595 -0.003890  0.024438           2  ...  0.141703   \n",
      "\n",
      "        symbol                                          name  \\\n",
      "0       NDUFA4       NDUFA4 mitochondrial complex associated   \n",
      "1       NDUFA4       NDUFA4 mitochondrial complex associated   \n",
      "2       NDUFA4       NDUFA4 mitochondrial complex associated   \n",
      "3       NDUFA4       NDUFA4 mitochondrial complex associated   \n",
      "4       NDUFA4       NDUFA4 mitochondrial complex associated   \n",
      "...        ...                                           ...   \n",
      "621501   SCN7A  sodium voltage-gated channel alpha subunit 7   \n",
      "621502   SCN7A  sodium voltage-gated channel alpha subunit 7   \n",
      "621503   SCN7A  sodium voltage-gated channel alpha subunit 7   \n",
      "621504   SCN7A  sodium voltage-gated channel alpha subunit 7   \n",
      "621505   SCN7A  sodium voltage-gated channel alpha subunit 7   \n",
      "\n",
      "                                                    alias    type_of_gene  \\\n",
      "0       [CI-9k, CI-MLRQ, COXFA4, MC4DN21, MISTR1, MLRQ...  protein-coding   \n",
      "1       [CI-9k, CI-MLRQ, COXFA4, MC4DN21, MISTR1, MLRQ...  protein-coding   \n",
      "2       [CI-9k, CI-MLRQ, COXFA4, MC4DN21, MISTR1, MLRQ...  protein-coding   \n",
      "3       [CI-9k, CI-MLRQ, COXFA4, MC4DN21, MISTR1, MLRQ...  protein-coding   \n",
      "4       [CI-9k, CI-MLRQ, COXFA4, MC4DN21, MISTR1, MLRQ...  protein-coding   \n",
      "...                                                   ...             ...   \n",
      "621501                       [NaG, Nav2.1, Nav2.2, SCN6A]  protein-coding   \n",
      "621502                       [NaG, Nav2.1, Nav2.2, SCN6A]  protein-coding   \n",
      "621503                       [NaG, Nav2.1, Nav2.2, SCN6A]  protein-coding   \n",
      "621504                       [NaG, Nav2.1, Nav2.2, SCN6A]  protein-coding   \n",
      "621505                       [NaG, Nav2.1, Nav2.2, SCN6A]  protein-coding   \n",
      "\n",
      "       map_location                                        genomic_pos  \\\n",
      "0            7p21.3  {'chr': '7', 'end': 10940153, 'ensemblgene': '...   \n",
      "1            7p21.3  {'chr': '7', 'end': 10940153, 'ensemblgene': '...   \n",
      "2            7p21.3  {'chr': '7', 'end': 10940153, 'ensemblgene': '...   \n",
      "3            7p21.3  {'chr': '7', 'end': 10940153, 'ensemblgene': '...   \n",
      "4            7p21.3  {'chr': '7', 'end': 10940153, 'ensemblgene': '...   \n",
      "...             ...                                                ...   \n",
      "621501       2q24.3  {'chr': '2', 'end': 166611482, 'ensemblgene': ...   \n",
      "621502       2q24.3  {'chr': '2', 'end': 166611482, 'ensemblgene': ...   \n",
      "621503       2q24.3  {'chr': '2', 'end': 166611482, 'ensemblgene': ...   \n",
      "621504       2q24.3  {'chr': '2', 'end': 166611482, 'ensemblgene': ...   \n",
      "621505       2q24.3  {'chr': '2', 'end': 166611482, 'ensemblgene': ...   \n",
      "\n",
      "        entrezgene   HGNC                                            summary  \n",
      "0             4697   7687  The protein encoded by this gene belongs to th...  \n",
      "1             4697   7687  The protein encoded by this gene belongs to th...  \n",
      "2             4697   7687  The protein encoded by this gene belongs to th...  \n",
      "3             4697   7687  The protein encoded by this gene belongs to th...  \n",
      "4             4697   7687  The protein encoded by this gene belongs to th...  \n",
      "...            ...    ...                                                ...  \n",
      "621501        6332  10594  This gene encodes one of the many voltage-gate...  \n",
      "621502        6332  10594  This gene encodes one of the many voltage-gate...  \n",
      "621503        6332  10594  This gene encodes one of the many voltage-gate...  \n",
      "621504        6332  10594  This gene encodes one of the many voltage-gate...  \n",
      "621505        6332  10594  This gene encodes one of the many voltage-gate...  \n",
      "\n",
      "[621506 rows x 34 columns]\n"
     ]
    }
   ],
   "source": [
    "# Merge the original pandas DataFrame (filtered_results_pd) with the gene information DataFrame on 'EnsemblID'\n",
    "filtered_results_annotated = filtered_results_pd.merge(gene_info_df, left_on='EnsemblID', right_on='query', how='left')\n",
    "\n",
    "# Drop the 'query' column (since it's the same as 'EnsemblID')\n",
    "filtered_results_annotated.drop(columns=['query'], inplace=True)\n",
    "\n",
    "# Display the annotated DataFrame\n",
    "print(filtered_results_annotated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows in filtered_results_annotated_pd: 621506\n"
     ]
    }
   ],
   "source": [
    "# Count the number of rows in filtered_results_annotated_pd\n",
    "num_rows = filtered_results_annotated.shape[0]\n",
    "\n",
    "# Print the number of rows\n",
    "print(f\"Number of rows in filtered_results_annotated_pd: {num_rows}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['EnsemblID', 'VariantID', 'tss_distance', 'CAF_eQTL', 'ma_samples',\n",
      "       'ma_count', 'pval_nominal', 'Beta', 'SE', 'chromosome', 'position',\n",
      "       'OtherAlleleA', 'CodedAlleleB', 'CAF', 'AltID', 'Source',\n",
      "       'AverageMaximumPosteriorCall', 'Info', 'AA_N', 'AB_N', 'BB_N', 'TotalN',\n",
      "       'MAF', 'MissingDataProportion', 'HWE_P', 'symbol', 'name', 'alias',\n",
      "       'type_of_gene', 'map_location', 'genomic_pos', 'entrezgene', 'HGNC',\n",
      "       'summary'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# Display the headers of the DataFrame\n",
    "print(filtered_results_annotated.columns)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Saving\n",
    "\n",
    "Let's save the annotated table.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the order of columns if needed\n",
    "columns_order = ['EnsemblID', 'symbol', \n",
    "'VariantID', 'tss_distance', 'CAF_eQTL', 'ma_samples',\n",
    "'ma_count', 'pval_nominal', 'Beta', 'SE', 'chromosome', 'position',\n",
    "'OtherAlleleA', 'CodedAlleleB', 'CAF', 'AltID', 'Source',\n",
    "'AverageMaximumPosteriorCall', 'Info', 'AA_N', 'AB_N', 'BB_N', 'TotalN',\n",
    "'MAF', 'MissingDataProportion', 'HWE_P', \n",
    "'name', 'alias', 'type_of_gene', 'map_location', 'genomic_pos', 'entrezgene', 'HGNC', 'summary']\n",
    "# Ensure these columns exist in your DataFrame\n",
    "\n",
    "# Save the DataFrame with specified columns\n",
    "filtered_results_annotated.to_csv(\n",
    "    os.path.join(\n",
    "        molQTL_loc, 'target_regions.nom_cis_eqtl_annotated.txt.gz'),\n",
    "    sep='\\t',\n",
    "    index=False,\n",
    "    compression='gzip',\n",
    "    columns=columns_order\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "del filtered_results # remove the filtered_results DataFrame from memory\n",
    "del filtered_results_annotated # remove the filtered_results_annotated DataFrame from memory\n",
    "del filtered # remove the filtered DataFrame from memory\n",
    "del filtered_results_pd # remove the filtered_results_pd DataFrame from memory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract _cis_-mQTL data\n",
    "\n",
    "Now we're ready to extract the data for each region in one table.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape: (268_375, 20)\n",
      "┌────────────┬─────────┬────────────┬────────────┬───┬──────────┬───────────┬───────────┬──────────┐\n",
      "│ phenotype_ ┆ num_var ┆ beta_shape ┆ beta_shape ┆ … ┆ qval     ┆ pval_nomi ┆ chromosom ┆ position │\n",
      "│ id         ┆ ---     ┆ 1          ┆ 2          ┆   ┆ ---      ┆ nal_thres ┆ e         ┆ ---      │\n",
      "│ ---        ┆ i64     ┆ ---        ┆ ---        ┆   ┆ f64      ┆ hold      ┆ ---       ┆ i64      │\n",
      "│ str        ┆         ┆ f64        ┆ f64        ┆   ┆          ┆ ---       ┆ str       ┆          │\n",
      "│            ┆         ┆            ┆            ┆   ┆          ┆ f64       ┆           ┆          │\n",
      "╞════════════╪═════════╪════════════╪════════════╪═══╪══════════╪═══════════╪═══════════╪══════════╡\n",
      "│ cg21870274 ┆ 309     ┆ 1.01037    ┆ 15.0882    ┆ … ┆ 0.468602 ┆ 0.001232  ┆ 1         ┆ 752307   │\n",
      "│ cg08258224 ┆ 731     ┆ 1.03658    ┆ 40.9219    ┆ … ┆ 0.605861 ┆ 0.000509  ┆ 1         ┆ 1636400  │\n",
      "│ cg18147296 ┆ 750     ┆ 1.02594    ┆ 43.8044    ┆ … ┆ 0.006203 ┆ 0.000454  ┆ 1         ┆ 771410   │\n",
      "│ cg13938959 ┆ 787     ┆ 1.00503    ┆ 41.991     ┆ … ┆ 0.45302  ┆ 0.000433  ┆ 1         ┆ 800193   │\n",
      "│ cg12445832 ┆ 787     ┆ 1.01013    ┆ 42.2714    ┆ … ┆ 0.211842 ┆ 0.000439  ┆ 1         ┆ 1706886  │\n",
      "│ …          ┆ …       ┆ …          ┆ …          ┆ … ┆ …        ┆ …         ┆ …         ┆ …        │\n",
      "│ cg19004771 ┆ 2979    ┆ 1.04971    ┆ 202.825    ┆ … ┆ 0.152963 ┆ 0.000135  ┆ 22        ┆ 51181759 │\n",
      "│ cg20569369 ┆ 2979    ┆ 1.04484    ┆ 216.682    ┆ … ┆ 0.249547 ┆ 0.000124  ┆ 22        ┆ 50519640 │\n",
      "│ cg26034629 ┆ 2975    ┆ 1.03817    ┆ 214.838    ┆ … ┆ 0.423997 ┆ 0.000122  ┆ 22        ┆ 50913182 │\n",
      "│ cg25232725 ┆ 2975    ┆ 1.04781    ┆ 217.38     ┆ … ┆ 0.300416 ┆ 0.000125  ┆ 22        ┆ 50642777 │\n",
      "│ cg05615487 ┆ 2973    ┆ 1.05074    ┆ 219.59     ┆ … ┆ 0.445053 ┆ 0.000126  ┆ 22        ┆ 51151350 │\n",
      "└────────────┴─────────┴────────────┴────────────┴───┴──────────┴───────────┴───────────┴──────────┘\n"
     ]
    }
   ],
   "source": [
    "# Split variant_id into chromosome and position directly\n",
    "sumstats_nom_cis_mqtl = sumstats_nom_cis_mqtl.with_columns([\n",
    "    pl.col(\"variant_id\"),  # Retain the original column\n",
    "    pl.col(\"variant_id\").str.extract(r\"(\\d+):\", 1).alias(\"chromosome\"),\n",
    "    pl.col(\"variant_id\").str.extract(r\":(\\d+)\", 1).cast(pl.Int64).alias(\"position\")\n",
    "])\n",
    "\n",
    "# Show the resulting DataFrame\n",
    "print(sumstats_nom_cis_mqtl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape: (3_558, 20)\n",
      "┌────────────┬─────────┬────────────┬───────────┬───┬──────────┬───────────┬───────────┬───────────┐\n",
      "│ phenotype_ ┆ num_var ┆ beta_shape ┆ beta_shap ┆ … ┆ qval     ┆ pval_nomi ┆ chromosom ┆ position  │\n",
      "│ id         ┆ ---     ┆ 1          ┆ e2        ┆   ┆ ---      ┆ nal_thres ┆ e         ┆ ---       │\n",
      "│ ---        ┆ i64     ┆ ---        ┆ ---       ┆   ┆ f64      ┆ hold      ┆ ---       ┆ i64       │\n",
      "│ str        ┆         ┆ f64        ┆ f64       ┆   ┆          ┆ ---       ┆ str       ┆           │\n",
      "│            ┆         ┆            ┆           ┆   ┆          ┆ f64       ┆           ┆           │\n",
      "╞════════════╪═════════╪════════════╪═══════════╪═══╪══════════╪═══════════╪═══════════╪═══════════╡\n",
      "│ cg10940374 ┆ 8142    ┆ 1.05747    ┆ 430.204   ┆ … ┆ 0.108641 ┆ 0.000057  ┆ 7         ┆ 11621877  │\n",
      "│ cg23711939 ┆ 8140    ┆ 1.05286    ┆ 415.566   ┆ … ┆ 0.234493 ┆ 0.000058  ┆ 7         ┆ 11902812  │\n",
      "│ cg23083824 ┆ 8086    ┆ 1.05272    ┆ 419.446   ┆ … ┆ 0.353023 ┆ 0.000058  ┆ 7         ┆ 11526379  │\n",
      "│ cg07287563 ┆ 8085    ┆ 1.06727    ┆ 369.628   ┆ … ┆ 0.595796 ┆ 0.000069  ┆ 7         ┆ 11382366  │\n",
      "│ cg22477374 ┆ 8085    ┆ 1.03771    ┆ 391.77    ┆ … ┆ 0.325596 ┆ 0.000058  ┆ 7         ┆ 11521941  │\n",
      "│ …          ┆ …       ┆ …          ┆ …         ┆ … ┆ …        ┆ …         ┆ …         ┆ …         │\n",
      "│ cg19049344 ┆ 4736    ┆ 1.06155    ┆ 257.846   ┆ … ┆ 0.379259 ┆ 0.000105  ┆ 2         ┆ 165648384 │\n",
      "│ cg00508771 ┆ 4751    ┆ 1.05287    ┆ 242.645   ┆ … ┆ 0.231803 ┆ 0.000107  ┆ 2         ┆ 165744679 │\n",
      "│ cg07322293 ┆ 5583    ┆ 1.04651    ┆ 261.41    ┆ … ┆ 0.398253 ┆ 0.000097  ┆ 2         ┆ 165761629 │\n",
      "│ cg11219022 ┆ 5459    ┆ 1.03068    ┆ 268.048   ┆ … ┆ 0.276384 ┆ 0.000089  ┆ 2         ┆ 165870133 │\n",
      "│ cg13268671 ┆ 5458    ┆ 1.04332    ┆ 254.938   ┆ … ┆ 0.543397 ┆ 0.000098  ┆ 2         ┆ 165858570 │\n",
      "└────────────┴─────────┴────────────┴───────────┴───┴──────────┴───────────┴───────────┴───────────┘\n"
     ]
    }
   ],
   "source": [
    "import polars as pl\n",
    "\n",
    "# Initialize an empty DataFrame with the same schema as sumstats_nom_cis_mqtl\n",
    "filtered_results = pl.DataFrame(schema=sumstats_nom_cis_mqtl.schema)\n",
    "\n",
    "# Loop through each row of df2 and filter df1 based on chromosome and position range\n",
    "for row in target_regions.iter_rows(named=True):\n",
    "    chrom = row['chromosome']\n",
    "    lower_pos = row['lower_position']\n",
    "    upper_pos = row['upper_position']\n",
    "    \n",
    "    # Filter df1 based on the chromosome and position range using polars\n",
    "    filtered = sumstats_nom_cis_mqtl.filter(\n",
    "        (sumstats_nom_cis_mqtl['chromosome'] == chrom) & \n",
    "        (sumstats_nom_cis_mqtl['position'] >= lower_pos) & \n",
    "        (sumstats_nom_cis_mqtl['position'] <= upper_pos)\n",
    "    )\n",
    "    \n",
    "    # Append the filtered rows to the result DataFrame\n",
    "    filtered_results = filtered_results.vstack(filtered)\n",
    "\n",
    "# Show the final filtered DataFrame\n",
    "print(filtered_results)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows in filtered_results: 3558\n"
     ]
    }
   ],
   "source": [
    "# Count the number of rows in the filtered_results DataFrame\n",
    "num_rows = filtered_results.shape[0]\n",
    "\n",
    "# Print the number of rows\n",
    "print(f\"Number of rows in filtered_results: {num_rows}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "del sumstats_nom_cis_mqtl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Annotation\n",
    "\n",
    "Let's also annotate this table with more information.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     phenotype_id  num_var  beta_shape1  beta_shape2  true_df  pval_true_df  \\\n",
      "0      cg10940374     8142      1.05747      430.204  329.506      0.000144   \n",
      "1      cg23711939     8140      1.05286      415.566  327.662      0.000447   \n",
      "2      cg23083824     8086      1.05272      419.446  329.860      0.000984   \n",
      "3      cg07287563     8085      1.06727      369.628  320.193      0.009988   \n",
      "4      cg22477374     8085      1.03771      391.770  325.199      0.000866   \n",
      "...           ...      ...          ...          ...      ...           ...   \n",
      "3553   cg19049344     4736      1.06155      257.846  330.619      0.002345   \n",
      "3554   cg00508771     4751      1.05287      242.645  327.344      0.000858   \n",
      "3555   cg07322293     5583      1.04651      261.410  328.283      0.002620   \n",
      "3556   cg11219022     5459      1.03068      268.048  332.278      0.001043   \n",
      "3557   cg13268671     5458      1.04332      254.938  328.934      0.013409   \n",
      "\n",
      "       variant_id  tss_distance  ma_samples  ma_count  ...  UCSC_RefGene_Name  \\\n",
      "0      7:11621877        641884          29        30  ...             NDUFA4   \n",
      "1      7:11902812        922533         282       332  ...             NDUFA4   \n",
      "2      7:11526379        513104          70        72  ...        PHF14;PHF14   \n",
      "3      7:11382366        369025         245       302  ...        PHF14;PHF14   \n",
      "4      7:11521941        508579         246       303  ...        PHF14;PHF14   \n",
      "...           ...           ...         ...       ...  ...                ...   \n",
      "3553  2:165648384       -447442          24        24  ...              SCN2A   \n",
      "3554  2:165744679       -408056         212       268  ...  SCN2A;SCN2A;SCN2A   \n",
      "3555  2:165761629       -864025         155       176  ...             GALNT3   \n",
      "3556  2:165870133       -940145         123       137  ...      TTC21B;TTC21B   \n",
      "3557  2:165858570       -952016          77        82  ...             TTC21B   \n",
      "\n",
      "                   UCSC_RefGene_Accession  UCSC_RefGene_Group  \\\n",
      "0                               NM_002489              TSS200   \n",
      "1                               NM_002489             TSS1500   \n",
      "2                  NM_001007157;NM_014660     TSS1500;TSS1500   \n",
      "3                  NM_001007157;NM_014660       TSS200;TSS200   \n",
      "4                  NM_001007157;NM_014660       TSS200;TSS200   \n",
      "...                                   ...                 ...   \n",
      "3553                         NM_001040142              TSS200   \n",
      "3554  NM_021007;NM_001040142;NM_001040143      Body;Body;Body   \n",
      "3555                            NM_004482                Body   \n",
      "3556                  NM_024753;NM_024753       1stExon;5'UTR   \n",
      "3557                            NM_024753             TSS1500   \n",
      "\n",
      "                           Phantom  DMR  Enhancer             HMM_Island  \\\n",
      "0                                                    7:10945952-10946804   \n",
      "1                                                    7:10945952-10946804   \n",
      "2                                                    7:10979797-10980088   \n",
      "3                                                    7:10979797-10980088   \n",
      "4                                                    7:10979797-10980088   \n",
      "...                            ...  ...       ...                    ...   \n",
      "3553                                                                       \n",
      "3554                                                                       \n",
      "3555                                                                       \n",
      "3556  high-CpG:166518504-166518595                 2:166518217-166519050   \n",
      "3557                                               2:166518217-166519050   \n",
      "\n",
      "      Regulatory_Feature_Name Regulatory_Feature_Group  DHS  \n",
      "0         7:10979259-10981086      Promoter_Associated       \n",
      "1         7:10979259-10981086      Promoter_Associated       \n",
      "2         7:11012661-11014378      Promoter_Associated       \n",
      "3         7:11012661-11014378      Promoter_Associated       \n",
      "4         7:11012661-11014378      Promoter_Associated       \n",
      "...                       ...                      ...  ...  \n",
      "3553                                                         \n",
      "3554                                                         \n",
      "3555                                                         \n",
      "3556    2:166809230-166811343      Promoter_Associated       \n",
      "3557    2:166809230-166811343      Promoter_Associated       \n",
      "\n",
      "[3558 rows x 53 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from rpy2.robjects import r, pandas2ri\n",
    "import rpy2.robjects.packages as rpackages\n",
    "\n",
    "# Enable the automatic conversion between R and pandas DataFrames\n",
    "pandas2ri.activate()\n",
    "\n",
    "# Load the Illumina annotation package in R\n",
    "illumina_pkg = rpackages.importr('IlluminaHumanMethylation450kanno.ilmn12.hg19')\n",
    "\n",
    "# Load and convert the annotation data to an R data frame\n",
    "r('library(IlluminaHumanMethylation450kanno.ilmn12.hg19)')\n",
    "r('annotation_df <- as.data.frame(getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19))')\n",
    "\n",
    "# Fetch the data frame into Python\n",
    "annotation_df = pandas2ri.rpy2py(r['annotation_df'])\n",
    "\n",
    "# Assuming 'phenotype_id' in filtered_results corresponds to CpG IDs in annotation_df\n",
    "filtered_results_pd = filtered_results.to_pandas()\n",
    "\n",
    "# Merge filtered_results with the annotation data\n",
    "filtered_results_annotated = filtered_results_pd.merge(annotation_df, left_on='phenotype_id', right_index=True, how='left')\n",
    "\n",
    "# Display the merged data\n",
    "print(filtered_results_annotated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows in filtered_results_annotated_pd: 3558\n"
     ]
    }
   ],
   "source": [
    "# Count the number of rows in filtered_results_annotated_pd\n",
    "num_rows = filtered_results_annotated.shape[0]\n",
    "\n",
    "# Print the number of rows\n",
    "print(f\"Number of rows in filtered_results_annotated_pd: {num_rows}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['phenotype_id', 'num_var', 'beta_shape1', 'beta_shape2', 'true_df',\n",
      "       'pval_true_df', 'variant_id', 'tss_distance', 'ma_samples', 'ma_count',\n",
      "       'af', 'pval_nominal', 'slope', 'slope_se', 'pval_perm', 'pval_beta',\n",
      "       'qval', 'pval_nominal_threshold', 'chromosome', 'position', 'chr',\n",
      "       'pos', 'strand', 'Name', 'AddressA', 'AddressB', 'ProbeSeqA',\n",
      "       'ProbeSeqB', 'Type', 'NextBase', 'Color', 'Probe_rs', 'Probe_maf',\n",
      "       'CpG_rs', 'CpG_maf', 'SBE_rs', 'SBE_maf', 'Islands_Name',\n",
      "       'Relation_to_Island', 'Forward_Sequence', 'SourceSeq', 'Random_Loci',\n",
      "       'Methyl27_Loci', 'UCSC_RefGene_Name', 'UCSC_RefGene_Accession',\n",
      "       'UCSC_RefGene_Group', 'Phantom', 'DMR', 'Enhancer', 'HMM_Island',\n",
      "       'Regulatory_Feature_Name', 'Regulatory_Feature_Group', 'DHS'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# Display the headers of the DataFrame\n",
    "print(filtered_results_annotated.columns)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Saving\n",
    "\n",
    "Let's save the annotated table.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the order of columns if needed\n",
    "columns_order = ['phenotype_id', 'num_var', 'beta_shape1', 'beta_shape2', 'true_df',\n",
    "       'pval_true_df', 'variant_id', 'chromosome', 'position', 'tss_distance', 'ma_samples', 'ma_count',\n",
    "       'af', 'pval_nominal', 'slope', 'slope_se', 'pval_perm', 'pval_beta',\n",
    "       'qval', 'pval_nominal_threshold', 'chr',\n",
    "       'pos', 'strand', 'Name', 'AddressA', 'AddressB', 'ProbeSeqA',\n",
    "       'ProbeSeqB', 'Type', 'NextBase', 'Color', 'Probe_rs', 'Probe_maf',\n",
    "       'CpG_rs', 'CpG_maf', 'SBE_rs', 'SBE_maf', 'Islands_Name',\n",
    "       'Relation_to_Island', 'Forward_Sequence', 'SourceSeq', 'Random_Loci',\n",
    "       'Methyl27_Loci', 'UCSC_RefGene_Name', 'UCSC_RefGene_Accession',\n",
    "       'UCSC_RefGene_Group', 'Phantom', 'DMR', 'Enhancer', 'HMM_Island',\n",
    "       'Regulatory_Feature_Name', 'Regulatory_Feature_Group', 'DHS']\n",
    "# Ensure these columns exist in your DataFrame\n",
    "\n",
    "# Save the DataFrame with specified columns\n",
    "filtered_results_annotated.to_csv(\n",
    "    os.path.join(\n",
    "        molQTL_loc, 'target_regions.nom_cis_mqtl_annotated.txt.gz'),\n",
    "    sep='\\t',\n",
    "    index=False,\n",
    "    compression='gzip',\n",
    "    columns=columns_order\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "del filtered_results # remove the filtered_results DataFrame from memory\n",
    "del filtered_results_annotated # remove the filtered_results_annotated DataFrame from memory\n",
    "del filtered # remove the filtered DataFrame from memory\n",
    "del filtered_results_pd # remove the filtered_results_pd DataFrame from memory\n",
    "del merged_data # remove the merged_data DataFrame from memory\n",
    "del annotation_data # remove the annotation_data DataFrame from memory"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gwaslab",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
